// @carbon/ibmdotcom-react/lib/components/FeatureCard/FeatureCard.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _Card = require("../Card");

var _classnames = _interopRequireDefault(require("classnames"));

var _settings = _interopRequireDefault(require("../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _settings2 = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var stablePrefix = _settings.default.stablePrefix;
var prefix = _settings2.default.prefix;
/**
 * Featured Card Component
 *
 * @param {object} props props object
 * @param {object} props.card section card object
 * @param {Function} props.onClick onClick function
 * @returns {*} FeatureCard JSX component
 */

var FeatureCard = function FeatureCard(_ref) {
  var _classNames, _classNames2;

  var card = _ref.card,
      size = _ref.size,
      otherProps = (0, _objectWithoutProperties2.default)(_ref, ["card", "size"]);
  var image = card.image,
      cta = card.cta,
      copy = card.copy;
  return cta && image && /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)((_classNames = {}, (0, _defineProperty2.default)(_classNames, "".concat(prefix, "--feature-card"), size == 'medium'), (0, _defineProperty2.default)(_classNames, "".concat(prefix, "--feature-card-large"), size == 'large'), (0, _defineProperty2.default)(_classNames, "".concat(prefix, "--feature-card-large_no-copy-text"), size === 'large' && !copy), _classNames)),
    "data-autoid": size === 'large' ? "".concat(stablePrefix, "--feature-card-large") : "".concat(stablePrefix, "--feature-card")
  }, /*#__PURE__*/_react.default.createElement(_Card.Card, (0, _extends2.default)({
    customClassName: (0, _classnames.default)((_classNames2 = {}, (0, _defineProperty2.default)(_classNames2, "".concat(prefix, "--feature-card__card"), size === 'medium'), (0, _defineProperty2.default)(_classNames2, "".concat(prefix, "--feature-card-large__card"), size === 'large'), _classNames2))
  }, card, otherProps)));
};

FeatureCard.propTypes = {
  /**
   * "Eyebrow" text above copy and CTA.
   */
  eyebrow: _propTypes.default.string,

  /**
   * Title of the Card item.
   */
  heading: _propTypes.default.string.isRequired,

  /**
   * Body text for the card.
   */
  copy: _propTypes.default.string,

  /**
   * Object containing Feature Card details.
   * In summary, has the following structure.
   *
   * | Name      | Data Type | Description                                                                                                                                                                        |
   * | --------- | --------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
   * | `heading` | String    | Title of the Card item.                                                                                                                                                            |
   * | `image`   | Object    | Image object used in the FeatureCard component. See [`<Image>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-image--default#props) for full usage details. |
   * | `cta`     | Object    | Object containing target and href of link. See below.                                                                                                                              |
   *
   * CTA object has the following structure in summary:
   *
   * | Name   | Data Type | Description                       |
   * | ------ | --------- | --------------------------------- |
   * | `href` | String    | Url of the FeatureCard component. |
   *
   * See [`<Card>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-card--static#props) for full usage details.
   */
  card: _propTypes.default.shape({
    heading: _propTypes.default.string,
    eyebrow: _propTypes.default.string,
    copy: _propTypes.default.string,
    cta: _propTypes.default.shape({
      copy: _propTypes.default.string,
      href: _propTypes.default.string,
      type: _propTypes.default.oneOf(['jump', 'local', 'external', 'download', 'video'])
    }),
    image: _propTypes.default.shape({
      classname: _propTypes.default.string,
      sources: _propTypes.default.arrayOf(_propTypes.default.shape({
        src: _propTypes.default.string,
        breakpoint: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number])
      })),
      defaultSrc: _propTypes.default.string.isRequired,
      alt: _propTypes.default.string.isRequired,
      longDescription: _propTypes.default.string
    }),
    inverse: _propTypes.default.bool,
    customClassName: _propTypes.default.string,
    type: _propTypes.default.oneOf(['link'])
  }).isRequired,

  /**
   * The handler for `onclick` event.
   */
  onClick: _propTypes.default.func,

  /**
   * Size of Feature Card. Choose from:
   *
   * | Name    | Description                                                  |
   * | ------- | -------------------------------------------------------------|
   * | `medium`| Default Feature Card variant                                 |
   * | `large` | Large Feature Card variant that contains eyebrow and heading |
   */
  size: _propTypes.default.oneOf(['medium', 'large'])
};
FeatureCard.defaultProps = {
  size: 'medium'
};
var _default = FeatureCard;
exports.default = _default;
