// @carbon/ibmdotcom-react/lib/components/ContentBlockHeadlines/ContentBlockHeadlines.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _ContentBlock = _interopRequireDefault(require("../ContentBlock/ContentBlock"));

var _CTA = require("../CTA");

var _FeatureFlags = require("../../internal/FeatureFlags");

var _settings = _interopRequireDefault(require("../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings"));

var _featureflag = _interopRequireDefault(require("../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/featureflag/featureflag"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _settings2 = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var stablePrefix = _settings.default.stablePrefix;
var prefix = _settings2.default.prefix;
/**
 * ContentBlockHeadlines pattern
 */

var ContentBlockHeadlines = function ContentBlockHeadlines(_ref) {
  var heading = _ref.heading,
      copy = _ref.copy,
      items = _ref.items;
  return (0, _featureflag.default)(_FeatureFlags.DDS_CONTENTBLOCK_HEADLINES, /*#__PURE__*/_react.default.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--content-block-headlines"),
    className: "".concat(prefix, "--content-block-headlines")
  }, /*#__PURE__*/_react.default.createElement(_ContentBlock.default, {
    heading: heading,
    copy: copy,
    border: true
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--content-block-headlines__container")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--content-block-headlines__row")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--content-block-headlines__item-container")
  }, renderItems(items)))))));
};
/**
 * Renders the ContentBlockHeadlines items
 *
 * @param {Array} items array of content rows
 * @private
 * @returns {*} JSX component
 */


function renderItems(items) {
  var headlineItems = [];
  items.forEach(function (item, index) {
    headlineItems.push( /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefix, "--content-block-headlines__item"),
      key: index
    }, /*#__PURE__*/_react.default.createElement("h4", {
      className: "".concat(prefix, "--content-block-headlines__heading")
    }, item.headline), /*#__PURE__*/_react.default.createElement("p", {
      className: "".concat(prefix, "--content-block-headlines__copy")
    }, item.copy), item.cta && /*#__PURE__*/_react.default.createElement(_CTA.CTA, (0, _extends2.default)({
      customClassName: "".concat(prefix, "--content-block-headlines__cta-container")
    }, item.cta))));
  });
  return headlineItems;
}

ContentBlockHeadlines.propTypes = {
  /**
   * Heading of the content block.
   */
  heading: _propTypes.default.string.isRequired,

  /**
   * Intro copy of the content block.
   */
  copy: _propTypes.default.string.isRequired,

  /**
   * Array of content items.
   * See [CTA](http://www.ibm.com/standards/carbon/react/?path=/docs/components-cta--default#text-link) for full usage details.
   */
  items: _propTypes.default.arrayOf(_propTypes.default.shape({
    headline: _propTypes.default.string.isRequired,
    copy: _propTypes.default.string.isRequired,
    cta: _propTypes.default.shape({
      type: _propTypes.default.oneOfType([_propTypes.default.oneOf(['local', 'download', 'jump', 'external', 'video', 'default']), _propTypes.default.arrayOf(_propTypes.default.oneOf(['local', 'download', 'jump', 'external', 'video', 'default']))]),
      copy: _propTypes.default.string,
      href: _propTypes.default.string,
      customClassName: _propTypes.default.string
    })
  })).isRequired
};
var _default = ContentBlockHeadlines;
exports.default = _default;
