// @carbon/ibmdotcom-react/lib/internal/vendor/carbon-components-react/components/DatePicker/next/DatePicker.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Skeleton = exports.SingleWithCalendarWithLayer = exports.SingleWithCalendar = exports.SimpleWithLayer = exports.Simple = exports.RangeWithCalendarWithLayer = exports.RangeWithCalendar = exports.DatePickerPlayground = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _addonActions = require("@storybook/addon-actions");

var _DatePicker = _interopRequireDefault(require("./DatePicker"));

var _DatePicker2 = _interopRequireDefault(require("../DatePicker.Skeleton"));

var _DatePickerInput = _interopRequireDefault(require("../../DatePickerInput"));

var _Layer = require("../../Layer");

/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
// const patterns = {
//   'Short (d{1,2}/d{4})': '\\d{1,2}/\\d{4}',
//   'Regular (d{1,2}/d{1,2}/d{4})': '\\d{1,2}/\\d{1,2}/\\d{4}',
// };
// const sizes = {
//   'Small  (sm)': 'sm',
//   'Medium (md) - default': undefined,
//   'Large  (lg)': 'lg',
// };
// const types = {
//   Simple: 'simple',
//   Single: 'single',
//   Range: 'range',
// };
var props = {
  datePicker: function datePicker() {
    return {
      dateFormat: 'm/d/Y',
      id: 'date-picker',
      light: false,
      datePickerType: 'single',
      minDate: '11/15/2020',
      maxDate: '11/01/2040',
      disable: ['07/15/2021'],
      onClose: (0, _addonActions.action)('onClose'),
      onOpen: (0, _addonActions.action)('onOpen')
    };
  },
  datePickerInput: function datePickerInput() {
    return {
      id: 'date-picker-input-id',
      className: 'some-class',
      size: undefined,
      labelText: 'Date Picker label',
      pattern: 'd{1,2}/d{4}',
      placeholder: 'mm/dd/yyyy',
      disabled: false,
      invalid: false,
      invalidText: 'A valid date is required',
      warn: false,
      warnText: 'Selected dates may cause conflicts',
      iconDescription: 'Icon description'
    };
  }
};
var _default = {
  title: 'Components/DatePicker',
  component: _DatePicker.default,
  subcomponents: {
    DatePickerInput: _DatePickerInput.default,
    DatePickerSkeleton: _DatePicker2.default
  }
};
exports.default = _default;

var Simple = function Simple() {
  return /*#__PURE__*/_react.default.createElement(_DatePicker.default, {
    datePickerType: "simple"
  }, /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    placeholder: "mm/dd/yyyy",
    labelText: "Date Picker label",
    id: "date-picker-simple"
  }));
};

exports.Simple = Simple;

var SingleWithCalendar = function SingleWithCalendar() {
  return /*#__PURE__*/_react.default.createElement(_DatePicker.default, {
    datePickerType: "single"
  }, /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    placeholder: "mm/dd/yyyy",
    labelText: "Date Picker label",
    id: "date-picker-single"
  }));
};

exports.SingleWithCalendar = SingleWithCalendar;

var RangeWithCalendar = function RangeWithCalendar() {
  return /*#__PURE__*/_react.default.createElement(_DatePicker.default, {
    datePickerType: "range"
  }, /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    id: "date-picker-input-id-start",
    placeholder: "mm/dd/yyyy",
    labelText: "Start date"
  }), /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    id: "date-picker-input-id-finish",
    placeholder: "mm/dd/yyyy",
    labelText: "End date"
  }));
};

exports.RangeWithCalendar = RangeWithCalendar;

var SimpleWithLayer = function SimpleWithLayer() {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_DatePicker.default, {
    datePickerType: "simple"
  }, /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    placeholder: "mm/dd/yyyy",
    labelText: "Date Picker label",
    id: "date-picker-simple"
  })), /*#__PURE__*/_react.default.createElement(_Layer.Layer, null, /*#__PURE__*/_react.default.createElement(_DatePicker.default, {
    datePickerType: "simple"
  }, /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    placeholder: "mm/dd/yyyy",
    labelText: "Date Picker label",
    id: "date-picker-simple"
  })), /*#__PURE__*/_react.default.createElement(_Layer.Layer, null, /*#__PURE__*/_react.default.createElement(_DatePicker.default, {
    datePickerType: "simple"
  }, /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    placeholder: "mm/dd/yyyy",
    labelText: "Date Picker label",
    id: "date-picker-simple"
  })))));
};

exports.SimpleWithLayer = SimpleWithLayer;

var SingleWithCalendarWithLayer = function SingleWithCalendarWithLayer() {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_DatePicker.default, {
    datePickerType: "single"
  }, /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    placeholder: "mm/dd/yyyy",
    labelText: "Date Picker label",
    id: "date-picker-single"
  })), /*#__PURE__*/_react.default.createElement(_Layer.Layer, null, /*#__PURE__*/_react.default.createElement(_DatePicker.default, {
    datePickerType: "single"
  }, /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    placeholder: "mm/dd/yyyy",
    labelText: "Date Picker label",
    id: "date-picker-single"
  })), /*#__PURE__*/_react.default.createElement(_Layer.Layer, null, /*#__PURE__*/_react.default.createElement(_DatePicker.default, {
    datePickerType: "single"
  }, /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    placeholder: "mm/dd/yyyy",
    labelText: "Date Picker label",
    id: "date-picker-single"
  })))));
};

exports.SingleWithCalendarWithLayer = SingleWithCalendarWithLayer;

var RangeWithCalendarWithLayer = function RangeWithCalendarWithLayer() {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_DatePicker.default, {
    datePickerType: "range"
  }, /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    id: "date-picker-input-id-start",
    placeholder: "mm/dd/yyyy",
    labelText: "Start date"
  }), /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    id: "date-picker-input-id-finish",
    placeholder: "mm/dd/yyyy",
    labelText: "End date"
  })), /*#__PURE__*/_react.default.createElement(_Layer.Layer, null, /*#__PURE__*/_react.default.createElement(_DatePicker.default, {
    datePickerType: "range"
  }, /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    id: "date-picker-input-id-start",
    placeholder: "mm/dd/yyyy",
    labelText: "Start date"
  }), /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    id: "date-picker-input-id-finish",
    placeholder: "mm/dd/yyyy",
    labelText: "End date"
  })), /*#__PURE__*/_react.default.createElement(_Layer.Layer, null, /*#__PURE__*/_react.default.createElement(_DatePicker.default, {
    datePickerType: "range"
  }, /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    id: "date-picker-input-id-start",
    placeholder: "mm/dd/yyyy",
    labelText: "Start date"
  }), /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, {
    id: "date-picker-input-id-finish",
    placeholder: "mm/dd/yyyy",
    labelText: "End date"
  })))));
};
/* eslint-disable react/prop-types */


exports.RangeWithCalendarWithLayer = RangeWithCalendarWithLayer;

var DatePickerPlayground = function DatePickerPlayground() {
  return /*#__PURE__*/_react.default.createElement(_DatePicker.default, props.datePicker(), /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, props.datePickerInput()), props.datePicker().datePickerType === 'range' && /*#__PURE__*/_react.default.createElement(_DatePickerInput.default, (0, _extends2.default)({}, props.datePickerInput(), {
    id: "date-picker-input-2"
  })));
};
/* eslint-enable react/prop-types */


exports.DatePickerPlayground = DatePickerPlayground;

var Skeleton = function Skeleton() {
  return /*#__PURE__*/_react.default.createElement(_DatePicker2.default, {
    range: true
  });
};

exports.Skeleton = Skeleton;
