// @carbon/ibmdotcom-react/es/components/ContentBlock/ContentBlock.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _extends from "@babel/runtime/helpers/extends";

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { CTA } from '../CTA';
import cx from 'classnames';
import ddsSettings from '../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings';
import { HorizontalRule } from '../HorizontalRule';
import markdownToHtml from '@carbon/ibmdotcom-utilities/es/utilities/markdownToHtml/markdownToHtml';
import PropTypes from 'prop-types';
import React from 'react';
import settings from 'carbon-components/es/globals/js/settings';
var stablePrefix = ddsSettings.stablePrefix;
var prefix = settings.prefix;
/**
 * ContentBlock internal component
 */

var ContentBlock = function ContentBlock(_ref) {
  var heading = _ref.heading,
      copy = _ref.copy,
      children = _ref.children,
      customClassName = _ref.customClassName,
      cta = _ref.cta,
      aside = _ref.aside,
      border = _ref.border;
  var classnames = cx("".concat(prefix, "--content-block"), customClassName);
  var content = /*#__PURE__*/React.createElement(React.Fragment, null, copy && /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--content-block__copy"),
    dangerouslySetInnerHTML: {
      __html: markdownToHtml(copy, {
        bold: false
      })
    }
  }), /*#__PURE__*/React.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--content-block__children"),
    className: "".concat(prefix, "--content-block__children")
  }, children), cta && _renderCTA(cta, border));
  var title = /*#__PURE__*/React.createElement(React.Fragment, null, heading && /*#__PURE__*/React.createElement("h2", {
    "data-autoid": "".concat(stablePrefix, "--content-block__heading"),
    className: "".concat(prefix, "--content-block__heading")
  }, heading));
  return /*#__PURE__*/React.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--content-block"),
    className: classnames
  }, title, aside && aside.items ? _layoutWrap(content, aside) : content, border ? /*#__PURE__*/React.createElement(HorizontalRule, null) : '');
};
/**
 * lays out the content with aside items in 2-1 format
 *
 * @private
 * @param {Element} content content elements
 * @param {Element} aside  aside elements
 * @returns {*} jsx cta component
 */


var _layoutWrap = function _layoutWrap(content, aside) {
  return /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--content-layout ").concat(prefix, "--content-layout--with-complementary")
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--content-layout__body")
  }, content), /*#__PURE__*/React.createElement("aside", {
    className: "".concat(prefix, "--content-layout__aside")
  }, aside.items));
};
/**
 * sets the class name based on theme type
 *
 * @private
 * @param {object} cta a cta object
 * @returns {*} jsx cta component
 */


function _renderCTA(cta, border) {
  if (cta.style === 'feature') {
    return /*#__PURE__*/React.createElement(CTA, _extends({
      customClassName: cx("".concat(prefix, "--content-block__cta"))
    }, cta));
  }

  return /*#__PURE__*/React.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--content-block__cta"),
    className: cx("".concat(prefix, "--content-block__cta-row"), _defineProperty({}, "".concat(prefix, "--content-block__cta-row-border"), border))
  }, /*#__PURE__*/React.createElement(CTA, _extends({
    customClassName: "".concat(prefix, "--content-block__cta ").concat(prefix, "--content-block__cta-col")
  }, cta)));
}

ContentBlock.propTypes = {
  /**
   * Heading text.
   */
  heading: PropTypes.string,

  /**
   * Copy text.
   */
  copy: PropTypes.string,

  /**
   * Children elements passed into `ContentBlock` to be rendered.
   */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),

  /**
   * CTA object.
   * See the [`<CTA>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-cta--default#props) for full usage details.
   */
  cta: PropTypes.shape({
    style: PropTypes.oneOf(['text', 'card', 'button', 'feature']),
    type: PropTypes.oneOfType([PropTypes.oneOf(['jump', 'local', 'external', 'download', 'video']), PropTypes.arrayOf(PropTypes.oneOf(['jump', 'local', 'external', 'download', 'video']))]),
    copy: PropTypes.string,
    href: PropTypes.string,
    customClassName: PropTypes.string
  }),

  /**
   * Custom className to wrap the `<ContentBlock>` component.
   */
  customClassName: PropTypes.string,

  /**
   * Object containing elements to be rendered within `<aside>` html element on right panel.
   * The structure is:
   *
   * | Name     | Data Type | Description                                                |
   * | -------- | --------- | ---------------------------------------------------------- |
   * | `items`  | Element   | Elements/Components to be rendered on the right panel.     |
   */
  aside: PropTypes.shape({
    items: PropTypes.element
  }),

  /**
   * border for content block.
   */
  border: PropTypes.bool
};
ContentBlock.defaultProps = {
  border: false
};
export default ContentBlock;
