// @carbon/ibmdotcom-react/lib/internal/vendor/@carbon/ibmdotcom-services/services/CloudAccountAuth/CloudAccountAuth.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _axios = _interopRequireDefault(require("axios"));

var _DDO = require("../DDO");

var _windowOrGlobal = _interopRequireDefault(require("window-or-global"));

/**
 * Copyright IBM Corp. 2021, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var CloudAccountAuthAPI = /*#__PURE__*/function () {
  function CloudAccountAuthAPI() {
    (0, _classCallCheck2.default)(this, CloudAccountAuthAPI);
  }

  (0, _createClass2.default)(CloudAccountAuthAPI, null, [{
    key: "checkPersonalization",

    /**
     * retrieve the cloud login status via window object status
     * gets the full digitalData (DDO) object.
     *
     * @example
     * import { cloudAccountAuthentication } from '@carbon/ibmdotcom-utilities';
     *
     * const status = cloudAccountAuthentication.checkPersonalization();
     *
     * @returns {string} string determining login status
     */
    value: function () {
      var _checkPersonalization = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return _DDO.DDOAPI.isReady().then(function () {
                  var _root$digitalData, _root$digitalData$use, _root$digitalData$use2;

                  var status = (_root$digitalData = _windowOrGlobal.default.digitalData) === null || _root$digitalData === void 0 ? void 0 : (_root$digitalData$use = _root$digitalData.user) === null || _root$digitalData$use === void 0 ? void 0 : (_root$digitalData$use2 = _root$digitalData$use.segment) === null || _root$digitalData$use2 === void 0 ? void 0 : _root$digitalData$use2.isCloudLoggedOn;
                  return {
                    user: status === true ? 'authenticated' : 'anonymous'
                  };
                });

              case 2:
                return _context.abrupt("return", _context.sent);

              case 3:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      function checkPersonalization() {
        return _checkPersonalization.apply(this, arguments);
      }

      return checkPersonalization;
    }()
    /**
     * retrieve the cloud login status via api
     *
     * @example
     * import { cloudAccountAuthentication } from '@carbon/ibmdotcom-utilities';
     *
     * const status = cloudAccountAuthentication.checkAPI();
     *
     * @returns {string} string determining login status
     */

  }, {
    key: "checkAPI",
    value: function () {
      var _checkAPI = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2() {
        var cloudLogin;
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return _axios.default.get('/api/v6/selected-account?profile=true', {
                  headers: {
                    Accept: 'application/json'
                  }
                }).then(function (response) {
                  return response.status === 200 ? 'authenticated' : 'anonymous';
                }).catch(function (error) {
                  console.error(error);
                  return 'anonymous';
                });

              case 2:
                cloudLogin = _context2.sent;
                return _context2.abrupt("return", {
                  user: cloudLogin
                });

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      }));

      function checkAPI() {
        return _checkAPI.apply(this, arguments);
      }

      return checkAPI;
    }()
  }]);
  return CloudAccountAuthAPI;
}();

var _default = CloudAccountAuthAPI;
exports.default = _default;
