// @carbon/ibmdotcom-react/es/components/DotcomShell/DotcomShell.js
/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import ddsSettings from '../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings';
import { Footer } from '../Footer';
import { Masthead } from '../Masthead';
import PropTypes from 'prop-types';
import React from 'react';
import settings from 'carbon-components/es/globals/js/settings';
var stablePrefix = ddsSettings.stablePrefix;
var prefix = settings.prefix;
/**
 * DotcomShell component.
 */

var DotcomShell = function DotcomShell(_ref) {
  var children = _ref.children,
      footerProps = _ref.footerProps,
      mastheadProps = _ref.mastheadProps;
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Masthead, mastheadProps), /*#__PURE__*/React.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--dotcom-shell"),
    className: "".concat(prefix, "--dotcom-shell")
  }, /*#__PURE__*/React.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--dotcom-shell__content"),
    className: "".concat(prefix, "--dotcom-shell__content")
  }, children)), /*#__PURE__*/React.createElement(Footer, footerProps));
};

DotcomShell.propTypes = {
  /**
   * Component(s) to render within the UI shell.
   */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,

  /**
   * Props for the Masthead.
   * See [`<Footer>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-footer--default) for more details.
   */
  footerProps: PropTypes.shape({
    navigation: PropTypes.shape({
      footerMenu: PropTypes.arrayOf(PropTypes.shape({
        title: PropTypes.string,
        links: PropTypes.arrayOf(PropTypes.shape({
          title: PropTypes.string,
          url: PropTypes.string
        }))
      })),
      footerThin: PropTypes.arrayOf(PropTypes.shape({
        title: PropTypes.string,
        url: PropTypes.string
      }))
    }),
    type: PropTypes.oneOf(['default', 'short', 'micro']),
    langCode: PropTypes.shape({
      cc: PropTypes.string,
      lc: PropTypes.string
    }),
    disableLocaleButton: PropTypes.bool,
    languageOnly: PropTypes.bool,
    languageItems: PropTypes.arrayOf(PropTypes.shape({})),
    languageInitialItem: PropTypes.shape({
      id: PropTypes.string,
      text: PropTypes.string
    }),
    languageCallback: PropTypes.func
  }),

  /**
   * Props for the Masthead.
   * See [`<Masthead>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-masthead--default) for more details.
   */
  mastheadProps: PropTypes.shape(Masthead.propTypes)
};
Footer.defaultProps = {
  footerProps: null,
  mastheadProps: null
};
export default DotcomShell;
