// @carbon/ibmdotcom-react/es/components/Masthead/HeaderNavContainer.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";

/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { keys, matches } from '../../internal/vendor/carbon-components-react/internal/keyboard';
import React, { useCallback, useEffect, useRef, useState } from 'react';
import CaretLeft20 from '@carbon/icons-react/es/caret--left/20';
import CaretRight20 from '@carbon/icons-react/es/caret--right/20';
import PropTypes from 'prop-types';
import root from 'window-or-global';
import settings from 'carbon-components/es/globals/js/settings';
var prefix = settings.prefix;
/**
 * Header nav container component.
 */

var HeaderNavContainer = function HeaderNavContainer(_ref) {
  var _root$document;

  var children = _ref.children;
  var containerRef = useRef(null);
  var contentRef = useRef(null);
  var contentLeftRef = useRef(null);
  var contentRightRef = useRef(null);
  var caretLeftRef = useRef(null);
  var caretRightRef = useRef(null);

  var _useState = useState(null),
      _useState2 = _slicedToArray(_useState, 2),
      io = _useState2[0],
      setIO = _useState2[1];

  var _useState3 = useState(0),
      _useState4 = _slicedToArray(_useState3, 2),
      position = _useState4[0],
      setPosition = _useState4[1];

  var buttonSize = 48; // 40px(width) + 8px(gradient)

  var pageIsRTL = ((_root$document = root.document) === null || _root$document === void 0 ? void 0 : _root$document.dir) === 'rtl';
  var paginateLeft = useCallback(function () {
    var menuItems = contentRef.current.querySelectorAll('.bx--header__menu-bar > li');

    if (pageIsRTL) {
      for (var i = 0; i < menuItems.length; i++) {
        if (contentRef.current.offsetWidth - menuItems[i].offsetLeft - buttonSize > -parseFloat(window.getComputedStyle(contentRef.current).right)) {
          setPosition(Math.min(-contentRef.current.offsetWidth + containerRef.current.offsetWidth + menuItems[i].offsetLeft - buttonSize, 0));
          contentRef.current.style.right = String(Math.min(-contentRef.current.offsetWidth + containerRef.current.offsetWidth + menuItems[i].offsetLeft - buttonSize, 0)) + 'px';
          break;
        }
      }
    } else {
      for (var _i = 0; _i < menuItems.length; _i++) {
        // checks if first visible item is partially hidden
        if (menuItems[_i].offsetLeft + menuItems[_i].offsetWidth + position >= buttonSize) {
          // checks if there is space for remaining menuItems
          if (menuItems[_i].offsetLeft + menuItems[_i].offsetWidth > containerRef.current.offsetWidth - buttonSize) {
            setPosition(containerRef.current.offsetWidth - menuItems[_i].offsetLeft - menuItems[_i].offsetWidth - buttonSize);
            contentRef.current.style.left = String(containerRef.current.offsetWidth - menuItems[_i].offsetLeft - menuItems[_i].offsetWidth - buttonSize) + 'px';
          } else {
            setPosition(0);
            contentRef.current.style.left = '0px';
          }

          break;
        }
      }
    }
  }, [position, pageIsRTL]);
  var paginateRight = useCallback(function () {
    var menuItems = contentRef.current.querySelectorAll('.bx--header__menu-bar > li');

    if (pageIsRTL) {
      for (var i = 0; i < menuItems.length; i++) {
        // checks if the right most visible element is partially hidden
        if (contentRef.current.offsetWidth - menuItems[i].offsetLeft > containerRef.current.offsetWidth - buttonSize - position) {
          setPosition(Math.max(-contentRef.current.offsetWidth + menuItems[i].offsetLeft + menuItems[i].offsetWidth + buttonSize, containerRef.current.offsetWidth - contentRef.current.offsetWidth));
          contentRef.current.style.right = String(Math.max(-contentRef.current.offsetWidth + menuItems[i].offsetLeft + menuItems[i].offsetWidth + buttonSize, containerRef.current.offsetWidth - contentRef.current.offsetWidth)) + 'px';
          break;
        }
      }
    } else {
      for (var _i2 = 0; _i2 < menuItems.length; _i2++) {
        if (menuItems[_i2].offsetLeft + menuItems[_i2].offsetWidth + position > containerRef.current.offsetWidth - buttonSize) {
          // checks if there is space for remaining menuItems
          if (contentRef.current.offsetWidth - menuItems[_i2].offsetLeft < containerRef.current.offsetWidth - buttonSize) {
            setPosition(containerRef.current.offsetWidth - contentRef.current.offsetWidth);
            contentRef.current.style.left = String(containerRef.current.offsetWidth - contentRef.current.offsetWidth) + 'px';
          } else {
            setPosition(buttonSize - menuItems[_i2].offsetLeft);
            contentRef.current.style.left = String(buttonSize - menuItems[_i2].offsetLeft) + 'px';
          }

          break;
        }
      }
    }
  }, [position, pageIsRTL]);
  useEffect(function () {
    if (window.IntersectionObserver) {
      setIO(new IntersectionObserver(function (records) {
        records.forEach(function (record) {
          if (contentLeftRef.current && record.target.classList.contains(contentLeftRef.current.className)) {
            caretLeftRef.current.hidden = record.isIntersecting;
          }

          if (contentRightRef.current && record.target.classList.contains(contentRightRef.current.className)) {
            caretRightRef.current.hidden = record.isIntersecting;
          }
        });
      }, {
        root: containerRef.current,
        threshold: 1
      }));
    }
  }, [setIO]);
  useEffect(function () {
    var navContent = contentRef.current;

    if (io) {
      navContent.addEventListener('keydown', handleOnKeyDown);
      io.observe(contentLeftRef.current);
      io.observe(contentRightRef.current);
    } else {
      return function () {
        if (io) {
          navContent.removeEventListener('keydown', handleOnKeyDown);
          io.disconnect();
        }
      };
    }
  });
  /**
   * Keyboard event handler for menu items.
   */

  var handleOnKeyDown = function handleOnKeyDown(event) {
    if (matches(event, [keys.Tab])) {
      if (pageIsRTL) {
        if (event.shiftKey) {
          var _document$activeEleme, _document$activeEleme2, _document$activeEleme3, _document$activeEleme4, _document$activeEleme5, _document$activeEleme6, _document$activeEleme7, _document$activeEleme8;

          if (((_document$activeEleme = document.activeElement) === null || _document$activeEleme === void 0 ? void 0 : (_document$activeEleme2 = _document$activeEleme.parentElement) === null || _document$activeEleme2 === void 0 ? void 0 : _document$activeEleme2.previousSibling) && ((_document$activeEleme3 = document.activeElement) === null || _document$activeEleme3 === void 0 ? void 0 : (_document$activeEleme4 = _document$activeEleme3.parentElement) === null || _document$activeEleme4 === void 0 ? void 0 : (_document$activeEleme5 = _document$activeEleme4.previousSibling) === null || _document$activeEleme5 === void 0 ? void 0 : _document$activeEleme5.offsetLeft) + ((_document$activeEleme6 = document.activeElement) === null || _document$activeEleme6 === void 0 ? void 0 : (_document$activeEleme7 = _document$activeEleme6.parentElement) === null || _document$activeEleme7 === void 0 ? void 0 : (_document$activeEleme8 = _document$activeEleme7.previousSibling) === null || _document$activeEleme8 === void 0 ? void 0 : _document$activeEleme8.offsetWidth) - parseFloat(window.getComputedStyle(contentRef.current).right) + buttonSize > contentRef.current.offsetWidth) {
            paginateLeft();
          }
        } else {
          var _document$activeEleme9, _document$activeEleme10, _document$activeEleme11, _document$activeEleme12, _document$activeEleme13;

          if (((_document$activeEleme9 = document.activeElement) === null || _document$activeEleme9 === void 0 ? void 0 : (_document$activeEleme10 = _document$activeEleme9.parentElement) === null || _document$activeEleme10 === void 0 ? void 0 : _document$activeEleme10.nextSibling) && contentRef.current.offsetWidth - ((_document$activeEleme11 = document.activeElement) === null || _document$activeEleme11 === void 0 ? void 0 : (_document$activeEleme12 = _document$activeEleme11.parentElement) === null || _document$activeEleme12 === void 0 ? void 0 : (_document$activeEleme13 = _document$activeEleme12.nextSibling) === null || _document$activeEleme13 === void 0 ? void 0 : _document$activeEleme13.offsetLeft) > containerRef.current.offsetWidth - buttonSize - position) {
            paginateRight();
          }
        }
      } else {
        if (event.shiftKey) {
          //Focus previous input
          if (document.activeElement.parentElement.previousSibling && document.activeElement.parentElement.previousSibling.offsetLeft + position <= buttonSize) {
            paginateLeft();
          }
        } else {
          //Focus next input
          if (document.activeElement.parentElement.nextSibling && document.activeElement.parentElement.nextSibling.offsetLeft + document.activeElement.parentElement.nextSibling.offsetWidth >= containerRef.current.offsetWidth - buttonSize) {
            paginateRight();
          }
        }
      }
    }
  };

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--header__nav-container"),
    ref: containerRef
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--header__nav-content"),
    ref: contentRef
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--sub-content-left"),
    ref: contentLeftRef
  }), /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--sub-content-right"),
    ref: contentRightRef
  }), children), /*#__PURE__*/React.createElement("div", {
    ref: caretLeftRef,
    className: "".concat(prefix, "--header__nav-caret-left-container"),
    hidden: true
  }, /*#__PURE__*/React.createElement("button", {
    className: "".concat(prefix, "--header__nav-caret-left"),
    "aria-label": "Masthead left caret",
    onClick: paginateLeft,
    tabIndex: "-1",
    "aria-hidden": "true"
  }, pageIsRTL ? /*#__PURE__*/React.createElement(CaretRight20, null) : /*#__PURE__*/React.createElement(CaretLeft20, null)), /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--header__nav-caret-left-gradient")
  })), /*#__PURE__*/React.createElement("div", {
    ref: caretRightRef,
    className: "".concat(prefix, "--header__nav-caret-right-container"),
    hidden: true
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--header__nav-caret-right-gradient")
  }), /*#__PURE__*/React.createElement("button", {
    className: "".concat(prefix, "--header__nav-caret-right"),
    "aria-label": "Masthead right caret",
    onClick: paginateRight,
    tabIndex: "-1",
    "aria-hidden": "true"
  }, pageIsRTL ? /*#__PURE__*/React.createElement(CaretLeft20, null) : /*#__PURE__*/React.createElement(CaretRight20, null)))));
};

HeaderNavContainer.propTypes = {
  /**
   * Container for other components.
   */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
};
export default HeaderNavContainer;
