// @carbon/ibmdotcom-react/es/components/Masthead/Masthead.js
import _extends from "@babel/runtime/helpers/extends";
import _typeof from "@babel/runtime/helpers/typeof";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { baseFontSize, breakpoints } from '@carbon/layout';
import React, { useCallback, useEffect, useRef, useState } from 'react';
import cx from 'classnames';
import { DDS_CUSTOM_PROFILE_LOGIN } from '../../internal/FeatureFlags';
import ddsSettings from '../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings';
import deprecate from '../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/deprecate/deprecate.js';
import { globalInit } from '../../internal/vendor/@carbon/ibmdotcom-services/services/global/global';
import Header from '../../internal/vendor/carbon-components-react/components/UIShell/Header';
import HeaderContainer from '../../internal/vendor/carbon-components-react/components/UIShell/HeaderContainer';
import HeaderGlobalBar from '../../internal/vendor/carbon-components-react/components/UIShell/HeaderGlobalBar';
import HeaderMenuButton from '../../internal/vendor/carbon-components-react/components/UIShell/HeaderMenuButton';
import { IbmLogo } from '../Icon';
import MastheadL1 from './MastheadL1';
import MastheadLeftNav from './MastheadLeftNav';
import MastheadProfile from './MastheadProfile';
import MastheadSearch from './MastheadSearch';
import MastheadTopNav from './MastheadTopNav';
import ProfileAPI from '../../internal/vendor/@carbon/ibmdotcom-services/services/Profile/Profile';
import PropTypes from 'prop-types';
import root from 'window-or-global';
import settings from 'carbon-components/es/globals/js/settings';
import SkipToContent from '../../internal/vendor/carbon-components-react/components/UIShell/SkipToContent';
import TranslationAPI from '../../internal/vendor/@carbon/ibmdotcom-services/services/Translation/Translation';
import User20 from '@carbon/icons-react/es/user/20';
import UserOnline20 from '@carbon/icons-react/es/user--online/20';
var stablePrefix = ddsSettings.stablePrefix;
var prefix = settings.prefix;
var gridBreakpoint = parseFloat(breakpoints.lg.width) * baseFontSize;
/**
 * MastHead component
 *
 * @param {object} props React props object
 * @param {object} props.navigation Object containing navigation elements
 * @param {boolean} props.hasProfile Determines whether to render Profile component
 * @param {boolean} props.hasSearch Determines whether to render Search Bar
 * @param {boolean} props.searchOpenOnload Determines if the search field is open on page load
 * @param {string} props.placeHolderText Placeholder value for search input
 * @param {string} props.initialSearchTerm Initial value for search input
 * @param {object} props.platform Platform name that appears on L0.
 * @param {string} props.title Title for the masthead L1
 * @param {string} props.eyebrowText Text for the eyebrow link in masthead L1
 * @param {string} props.eyebrowLink URL for the eyebrow link in masthead L1
 * @param {string} props.selectedMenuItem L0/L1 menu item to render with selected state
 * @returns {*} Masthead component
 */

var Masthead = function Masthead(_ref) {
  var _cx;

  var navigation = _ref.navigation,
      hasProfile = _ref.hasProfile,
      hasSearch = _ref.hasSearch,
      searchOpenOnload = _ref.searchOpenOnload,
      placeHolderText = _ref.placeHolderText,
      initialSearchTerm = _ref.initialSearchTerm,
      platform = _ref.platform,
      mastheadL1Data = _ref.mastheadL1Data,
      selectedMenuItem = _ref.selectedMenuItem,
      mastheadProps = _objectWithoutProperties(_ref, ["navigation", "hasProfile", "hasSearch", "searchOpenOnload", "placeHolderText", "initialSearchTerm", "platform", "mastheadL1Data", "selectedMenuItem"]);

  /**
   * Returns IBM.com authenticated status
   *
   * @param {boolean} isAuthenticated Whether the user is authenticated to IBM.com
   * @returns {*} The user status
   */
  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isAuthenticated = _useState2[0],
      setStatus = _useState2[1];
  /**
   * Returns state of search status
   *
   * @param {boolean} isSearchActive Whether the search bar is open
   * @returns {*} The active search status
   */


  var _useState3 = useState(searchOpenOnload),
      _useState4 = _slicedToArray(_useState3, 2),
      isSearchActive = _useState4[0],
      setIsSearchActive = _useState4[1];

  var searchIconButton = useRef(null);
  var handleChangeSearchActive = useCallback(function (event, _ref2) {
    var isOpen = _ref2.isOpen;
    setIsSearchActive(isOpen);
    setTimeout(function () {
      searchIconButton.current.focus();
    }, 0);
  }, []);
  useEffect(function () {
    // initialize global execution calls
    globalInit();
  }, []);
  useEffect(function () {
    var unmounted = false;

    _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
      var status;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return ProfileAPI.getUserStatus();

            case 2:
              status = _context.sent;

              if (!unmounted) {
                setStatus(status.user !== 'Unauthenticated');
              }

            case 4:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }))();

    return function () {
      unmounted = true;
    };
  }, []);

  var _useState5 = useState([]),
      _useState6 = _slicedToArray(_useState5, 2),
      mastheadData = _useState6[0],
      setMastheadData = _useState6[1];

  var _useState7 = useState({
    signedin: [],
    signedout: []
  }),
      _useState8 = _slicedToArray(_useState7, 2),
      profileData = _useState8[0],
      setProfileData = _useState8[1];

  useEffect(function () {
    var unmounted = false;

    _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
      var pageData;
      return _regeneratorRuntime.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.prev = 0;
              _context2.next = 3;
              return TranslationAPI.getTranslation();

            case 3:
              pageData = _context2.sent;

              if (!unmounted) {
                setMastheadData(pageData.mastheadNav.links);
                setProfileData(pageData.profileMenu);
              }

              _context2.next = 10;
              break;

            case 7:
              _context2.prev = 7;
              _context2.t0 = _context2["catch"](0);
              console.error('Error populating masthead data:', _context2.t0);

            case 10:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2, null, [[0, 7]]);
    }))();

    return function () {
      unmounted = true;
    };
  }, []);
  /**
   * Forces profile menu position to fixed to prevent scrolling
   *
   */

  var _setProfileListPosition = function _setProfileListPosition() {
    var profileMenuList = document.querySelector(".".concat(prefix, "--masthead__profile-item"));
    profileMenuList.closest('ul').style.position = 'fixed';
    profileMenuList.closest('ul').style.top = '48px';
  };

  var stickyRef = useRef(null);
  var mastheadL1Ref = useRef(null);
  var headerSearchClasses = cx((_cx = {}, _defineProperty(_cx, "".concat(prefix, "--masthead__platform"), platform), _defineProperty(_cx, "".concat(prefix, "--masthead__header--search-active"), isSearchActive), _cx));
  useEffect(function () {
    var tableOfContents = document.querySelector('.bx--tableofcontents__sidebar');
    var lastScrollPosition = 0;
    /**
     * Sets sticky masthead. If both L0 and L1 are present, L1 will be sticky.
     *
     */

    var handleScroll = root.addEventListener('scroll', function () {
      /**
       * L0 will hide on scroll down, show on scroll up
       *
       */
      if (mastheadL1Ref.current != null && tableOfContents != null) {
        var tocBoundingClient = tableOfContents.getBoundingClientRect();
        var mobileMastheadTop = Math.round(Math.min(0, tocBoundingClient.top - stickyRef.current.offsetHeight));
        var tocPosition = tocBoundingClient.top + lastScrollPosition - window.scrollY;
        tableOfContents.style.top = "".concat(Math.max(Math.min(tocPosition, stickyRef.current.offsetHeight), 0), "px");
        var regularMastheadTop = window.scrollY < lastScrollPosition ? 0 : -Math.min(stickyRef.current.offsetHeight - mastheadL1Ref.current.offsetHeight, Math.abs(mobileMastheadTop));
        var mastheadTop = window.innerWidth < gridBreakpoint ? mobileMastheadTop : regularMastheadTop;
        stickyRef.current.style.top = "".concat(mastheadTop, "px");
        stickyRef.current.style.transition = 'none';
        /**
         * L0 will hide on scroll down, show up on scroll up when mobile ToC is present
         */
      } else if (tableOfContents != null && stickyRef.current !== null && window.innerWidth < gridBreakpoint) {
        var _tocBoundingClient = tableOfContents.getBoundingClientRect();

        stickyRef.current.style.transition = "none";

        var _mastheadTop = Math.round(Math.min(0, _tocBoundingClient.top - stickyRef.current.offsetHeight));

        var _tocPosition = _tocBoundingClient.top + lastScrollPosition - window.scrollY;

        tableOfContents.style.top = "".concat(Math.max(Math.min(_tocPosition, stickyRef.current.offsetHeight), 0), "px");

        if (tableOfContents.style.top === '0px') {
          stickyRef.current.style.top = "-".concat(stickyRef.current.offsetHeight, "px");
        } else if (tableOfContents.style.top === "".concat(stickyRef.current.offsetHeight, "px")) {
          stickyRef.current.style.top = '0';
        } else {
          stickyRef.current.style.top = "".concat(_mastheadTop, "px");
        }
      }

      lastScrollPosition = window.scrollY;
    });
    return function () {
      root.removeEventListener('scroll', function () {
        return handleScroll;
      });
    };
  });

  if (navigation) {
    switch (_typeof(navigation)) {
      case 'default':
        // eslint-disable-next-line
        mastheadData = mastheadData;
        break;

      case 'object':
        mastheadData = navigation;
        break;

      default:
        break;
    }
  } // set navigation type (default, alternate, or ecosystem) for autoids


  var navType;

  if (!navigation && !platform) {
    navType = 'alt';
  } else if (navigation && !platform) {
    navType = 'default';
  } else if (platform) {
    navType = 'eco';
  }
  /**
   * checks if there is a child item in the menu section that matches current url and returns true for first valid result
   *
   * @returns {boolean} function that returns true or false
   */
  // eslint-disable-next-line class-methods-use-this


  var _hasCurrentUrl = function _hasCurrentUrl() {
    var matchFound = false;
    return function (sections, currentUrlPath) {
      if (!matchFound) {
        var _sections$menuSection;

        if (sections.url === currentUrlPath) {
          matchFound = true;
        } else if (sections === null || sections === void 0 ? void 0 : (_sections$menuSection = sections.menuSections) === null || _sections$menuSection === void 0 ? void 0 : _sections$menuSection[0]) {
          var _sections$menuSection2 = sections === null || sections === void 0 ? void 0 : sections.menuSections[0],
              menuItems = _sections$menuSection2.menuItems;

          for (var i = 0; i < menuItems.length; i++) {
            var _menuItems$i, _menuItems$i2, _menuItems$i2$megapan, _menuItems$i2$megapan2, _menuItems$i2$megapan3;

            if (((_menuItems$i = menuItems[i]) === null || _menuItems$i === void 0 ? void 0 : _menuItems$i.url) === currentUrlPath || ((_menuItems$i2 = menuItems[i]) === null || _menuItems$i2 === void 0 ? void 0 : (_menuItems$i2$megapan = _menuItems$i2.megapanelContent) === null || _menuItems$i2$megapan === void 0 ? void 0 : (_menuItems$i2$megapan2 = _menuItems$i2$megapan.quickLinks) === null || _menuItems$i2$megapan2 === void 0 ? void 0 : (_menuItems$i2$megapan3 = _menuItems$i2$megapan2.links) === null || _menuItems$i2$megapan3 === void 0 ? void 0 : _menuItems$i2$megapan3.filter(function (link) {
              return link.url === currentUrlPath;
            }).length)) {
              matchFound = true;
            }
          }
        }

        return matchFound;
      }

      return false;
    };
  };

  return /*#__PURE__*/React.createElement(HeaderContainer, {
    render: function render(_ref5) {
      var _mastheadL1Data$navig;

      var isSideNavExpanded = _ref5.isSideNavExpanded,
          onClickSideNavExpand = _ref5.onClickSideNavExpand;

      if (isSideNavExpanded) {
        var _root$document, _root$document$body;

        (_root$document = root.document) === null || _root$document === void 0 ? void 0 : (_root$document$body = _root$document.body) === null || _root$document$body === void 0 ? void 0 : _root$document$body.classList.add("".concat(prefix, "--body__lock-scroll"));
      } else {
        var _root$document2, _root$document2$body;

        (_root$document2 = root.document) === null || _root$document2 === void 0 ? void 0 : (_root$document2$body = _root$document2.body) === null || _root$document2$body === void 0 ? void 0 : _root$document2$body.classList.remove("".concat(prefix, "--body__lock-scroll"));
      }

      return /*#__PURE__*/React.createElement("div", {
        className: "".concat(prefix, "--masthead"),
        ref: stickyRef
      }, /*#__PURE__*/React.createElement("div", {
        className: "".concat(prefix, "--masthead__l0")
      }, /*#__PURE__*/React.createElement(Header, {
        "aria-label": "IBM",
        "data-autoid": "".concat(stablePrefix, "--masthead")
      }, /*#__PURE__*/React.createElement(SkipToContent, null), (mastheadL1Data || navigation) && /*#__PURE__*/React.createElement(HeaderMenuButton, {
        "aria-label": isSideNavExpanded ? 'Close menu' : 'Open menu',
        "data-autoid": "".concat(stablePrefix, "--masthead-").concat(navType, "-sidenav__l0-menu"),
        onClick: onClickSideNavExpand,
        isActive: isSideNavExpanded,
        className: headerSearchClasses,
        onBlur: function onBlur(e) {
          var platform = e.target.parentElement.querySelector("nav .".concat(prefix, "--side-nav__submenu-platform"));
          var firstMenuItem = e.target.parentElement.querySelector(".".concat(prefix, "--side-nav__menu-section--expanded li:first-of-type button")) || e.target.parentElement.querySelector(".".concat(prefix, "--side-nav__menu-section--expanded li:first-of-type a"));
          var lastMenuItem = e.target.parentElement.querySelector(".".concat(prefix, "--side-nav__menu-section--expanded li:last-of-type button")) || e.target.parentElement.querySelector(".".concat(prefix, "--side-nav__menu-section--expanded li:last-of-type a"));
          if (e.relatedTarget && e.relatedTarget !== firstMenuItem && e.relatedTarget !== platform) return lastMenuItem.focus();
        }
      }), (navigation || mastheadL1Data) && /*#__PURE__*/React.createElement(MastheadLeftNav, _extends({}, mastheadProps, {
        backButtonText: "Back",
        platform: platform,
        hasL1Data: !!mastheadL1Data,
        navigation: (_mastheadL1Data$navig = mastheadL1Data === null || mastheadL1Data === void 0 ? void 0 : mastheadL1Data.navigationL1) !== null && _mastheadL1Data$navig !== void 0 ? _mastheadL1Data$navig : mastheadData,
        isSideNavExpanded: isSideNavExpanded,
        navType: navType,
        selectedMenuItem: selectedMenuItem,
        onOverlayClick: onClickSideNavExpand
      })), /*#__PURE__*/React.createElement(IbmLogo, {
        logoData: mastheadProps.mastheadLogo,
        autoid: "".concat(stablePrefix, "--masthead-").concat(navType, "__l0-logo"),
        isSearchActive: isSearchActive
      }), /*#__PURE__*/React.createElement("div", {
        className: "".concat(prefix, "--header__search ").concat(headerSearchClasses)
      }, navigation && !mastheadL1Data && /*#__PURE__*/React.createElement(MastheadTopNav, _extends({}, mastheadProps, {
        platform: platform,
        navigation: mastheadData,
        navType: navType,
        hasCurrentUrl: _hasCurrentUrl,
        selectedMenuItem: selectedMenuItem
      })), hasSearch && /*#__PURE__*/React.createElement(MastheadSearch, _extends({}, mastheadProps, searchOpenOnload ? {
        searchOpenOnload: searchOpenOnload
      } : {}, {
        placeHolderText: placeHolderText,
        initialSearchTerm: initialSearchTerm,
        navType: navType,
        isSearchActive: isSearchActive,
        ref: searchIconButton,
        onChangeSearchActive: handleChangeSearchActive
      }))), hasProfile && /*#__PURE__*/React.createElement(HeaderGlobalBar, {
        className: "".concat(prefix, "--header__profile")
      }, /*#__PURE__*/React.createElement(MastheadProfile, _extends({
        overflowMenuProps: {
          ariaLabel: 'User Profile',
          'data-autoid': "".concat(stablePrefix, "--masthead-").concat(navType, "__l0-account"),
          flipped: true,
          style: {
            width: '3rem'
          },
          onOpen: function onOpen() {
            return _setProfileListPosition();
          },
          renderIcon: function renderIcon() {
            return isAuthenticated ? /*#__PURE__*/React.createElement(UserOnline20, null) : /*#__PURE__*/React.createElement(User20, null);
          }
        },
        overflowMenuItemProps: {
          wrapperClassName: "".concat(prefix, "--masthead__profile-item")
        },
        profileMenu: isAuthenticated ? profileData.signedin : profileData.signedout
      }, mastheadProps.customProfileLogin && DDS_CUSTOM_PROFILE_LOGIN ? {
        customProfileLogin: mastheadProps.customProfileLogin
      } : {}, {
        navType: navType
      }))))), mastheadL1Data && /*#__PURE__*/React.createElement("div", {
        ref: mastheadL1Ref
      }, /*#__PURE__*/React.createElement(MastheadL1, _extends({}, mastheadL1Data, {
        platform: platform,
        navType: navType,
        hasCurrentUrl: _hasCurrentUrl,
        selectedMenuItem: selectedMenuItem
      }))));
    }
  });
};

Masthead.propTypes = {
  /**
   * Navigation data object/string for Masthead. These navigation properties belongs to the Masthead L0 Top navigation. Use one from below:
   *
   * | Behavior           | Data Type | Description                                 | Example                             |
   * | ------------------ | --------- | ------------------------------------------- | ----------------------------------- |
   * | default navigation | String    | Default navigation data from IBM.com        | `<Masthead navigation="default" />` |
   * | custom navigation  | Object    | Pass in custom navigation data as an object | `<Masthead navigation={myNavObj}/>` |
   * | none               | null      | No navigation                               | `<Masthead />`                      |
   *
   * `Custom` navigation data must follow the same structure and key names as `default`.
   * See [this](https://1.www.s81c.com/common/carbon-for-ibm-dotcom/translations/masthead-footer/usen.json) for an example.
   */
  navigation: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.shape({
    hasMenu: PropTypes.bool,
    title: PropTypes.string,
    url: PropTypes.string,
    menuSections: PropTypes.arrayOf(PropTypes.shape({
      menuItems: PropTypes.arrayOf(PropTypes.shape({
        title: PropTypes.string,
        url: PropTypes.string
      }))
    }))
  }))]),

  /**
   * `true` to render IBM Profile Menu component.
   */
  hasProfile: PropTypes.bool,

  /**
   * Custom login url in masthead profile menu (experimental)
   */
  customProfileLogin: PropTypes.string,

  /**
   * `true` to render SearchBar component.
   */
  hasSearch: PropTypes.bool,

  /**
   * `true` to have search field open on page load. Does not close `onBlur`
   */
  searchOpenOnload: PropTypes.bool,

  /**
   * Platform name that appears on L0.
   * Includes platform name (only available with `default` and `custom navigation`).
   * Object requires `name` and `url`.
   * See [docs](http://www.ibm.com/standards/carbon/react/?path=/docs/components-masthead--default#platform) for more details.
   */
  platform: PropTypes.shape({
    name: PropTypes.string,
    url: PropTypes.string
  }),

  /**
   * L0 menu item to render with selected state. Needs to match `titleEnglish` field from nav data.
   */
  selectedMenuItem: PropTypes.string,

  /**
   * Placeholder value for search input.
   */
  placeHolderText: PropTypes.string,

  /**
   * Initial value for search input.
   */
  initialSearchTerm: PropTypes.string,

  /**
   * All the data that goes to the L1 of the Masthead.
   */
  mastheadL1Data: PropTypes.shape({
    /**
     * Platform name that appears on L1.
     * Includes platform name
     * Object requires `name` and `url`.
     * See [docs](http://www.ibm.com/standards/carbon/react/?path=/docs/components-masthead--default#platform) for more details.
     */
    platform: PropTypes.shape({
      name: PropTypes.string,
      url: PropTypes.string
    }),

    /**
     * Text for the eyebrow link in masthead L1 (experimental).
     */
    eyebrowText: PropTypes.string,

    /**
     * URL for the eyebrow link in masthead L1 (experimental).
     */
    eyebrowLink: PropTypes.string,

    /**
     * Navigation data object/string for Masthead L1. Use one from below:
     *
     * | Behavior           | Data Type | Description                                 | Example                             |
     * | ------------------ | --------- | ------------------------------------------- | ----------------------------------- |
     * | default navigation | String    | Default navigation data from IBM.com        | `<MastheadL1 navigationL1="default" />` |
     * | custom navigation  | Object    | Pass in custom navigation data as an object | `<MastheadL1 navigationL1={myNavObj}/>` |
     * | none               | null      | No navigation                               | `<MastheadL1 />`                      |
     *
     * `Custom` navigation data must follow the same structure and key names as `default`.
     * See [this](https://1.www.s81c.com/common/carbon-for-ibm-dotcom/translations/masthead-footer/usen.json) for an example.
     */
    navigationL1: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.shape({
      hasMenu: PropTypes.bool,
      title: PropTypes.string,
      url: PropTypes.string,
      menuSections: PropTypes.arrayOf(PropTypes.shape({
        menuItems: PropTypes.arrayOf(PropTypes.shape({
          title: PropTypes.string,
          url: PropTypes.string
        }))
      }))
    }))])
  }),

  /**
   * Custom typeahead API function
   */
  customTypeaheadApi: PropTypes.func,

  /**
   * Multiple search sections
   */
  multiSection: PropTypes.bool
};
Masthead.defaultProps = {
  hasProfile: true,
  hasSearch: true,
  searchOpenOnload: false,
  selectedMenuItem: '',
  platform: null,
  placeHolderText: 'Search all of IBM',
  initialSearchTerm: '',
  mastheadL1Data: null
};
export default deprecate(Masthead, "\n  The React Masthead is now in feature freeze; any new features or enhancements will only be added to the Web Components Masthead. \n  \n  We will continue to address any bugs specific to the React version for the rest of of Carbon for IBM.com v1.\n  \n  Please refer to the Web Components Masthead documentation for further details.\n  https://www.ibm.com/standards/carbon/web-components/?path=/docs/components-masthead--default\n");
