// @carbon/ibmdotcom-react/es/components/Masthead/MastheadL1.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React, { useEffect } from 'react';
import cx from 'classnames';
import ddsSettings from '../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings';
import HeaderMenu from '../carbon-components-react/UIShell/HeaderMenu';
import HeaderMenuItem from '../../internal/vendor/carbon-components-react/components/UIShell/HeaderMenuItem';
import HeaderNavContainer from './HeaderNavContainer';
import HeaderNavigation from '../../internal/vendor/carbon-components-react/components/UIShell/HeaderNavigation';
import MegaMenu from './MastheadMegaMenu/MegaMenu';
import PropTypes from 'prop-types';
import root from 'window-or-global';
import settings from 'carbon-components/es/globals/js/settings';
var stablePrefix = ddsSettings.stablePrefix;
var prefix = settings.prefix;
/**
 * MastHead L1 component.
 */

var MastheadL1 = function MastheadL1(_ref) {
  var navigationL1 = _ref.navigationL1,
      rest = _objectWithoutProperties(_ref, ["navigationL1"]);

  var className = cx(_defineProperty({}, "".concat(prefix, "--masthead__l1"), true));
  var childLinkChecker = rest.hasCurrentUrl();
  useEffect(function () {
    var _document$querySelect;

    (_document$querySelect = document.querySelector(".".concat(prefix, "--header__menu-bar"))) === null || _document$querySelect === void 0 ? void 0 : _document$querySelect.setAttribute('role', 'menu');
    document.querySelectorAll(".".concat(prefix, "--header__menu-bar li")).forEach(function (e) {
      e.setAttribute('role', 'menuitem');
      e.querySelector('a').removeAttribute('role');
    });
  }, []);
  var mastheadL1Links = navigationL1.map(function (link, index) {
    var _root$location;

    var selectedUrlItem = childLinkChecker && childLinkChecker(link, (_root$location = root.location) === null || _root$location === void 0 ? void 0 : _root$location.href);
    var autoid = "".concat(stablePrefix, "--masthead-").concat(rest.navType, "__l1-nav").concat(index);
    var selected = rest.selectedMenuItem ? link.titleEnglish === rest.selectedMenuItem : selectedUrlItem;

    if (link.hasMenu || link.hasMegapanel) {
      return /*#__PURE__*/React.createElement(HeaderMenu, {
        "aria-label": link.title,
        menuLinkName: link.title,
        className: cx(_defineProperty({}, "".concat(prefix, "--masthead__megamenu__l1-nav"), link.hasMegapanel)),
        selected: selected,
        autoId: autoid,
        key: index
      }, renderNav(link, rest.navType, autoid));
    }

    return /*#__PURE__*/React.createElement(HeaderMenuItem, {
      "data-selected": "".concat(!!selected),
      href: link.url,
      "data-autoid": autoid,
      key: index
    }, link.title);
  });
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", {
    className: className
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--masthead__l1-inner-container")
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--masthead__l1-name"),
    "data-selected": !rest.selectedMenuItem
  }, /*#__PURE__*/React.createElement("span", {
    className: "".concat(prefix, "--masthead__l1-name-title")
  }, /*#__PURE__*/React.createElement("a", {
    href: rest.platform.url
  }, rest.platform.name))), /*#__PURE__*/React.createElement(HeaderNavContainer, null, /*#__PURE__*/React.createElement(HeaderNavigation, {
    className: "".concat(prefix, "--masthead__l1-nav"),
    "aria-label": ""
  }, mastheadL1Links)))));
};
/**
 * Loops through and renders a list of links for the masthead nav
 *
 * @param {object} link A list of links to be rendered
 * @param {string} navType navigation type for autoids
 * @param {string} autoid autoid predecessor for megamenu items/menu items data-autoids
 * @returns {object} JSX object
 */


function renderNav(link, navType, autoid) {
  var navItems = [];

  if (link.hasMegapanel) {
    navItems.push( /*#__PURE__*/React.createElement(MegaMenu, {
      key: link.title,
      data: link,
      autoid: autoid
    }));
  } else {
    link.menuSections.forEach(function (section, i) {
      section.menuItems.forEach(function (item, j) {
        navItems.push( /*#__PURE__*/React.createElement(HeaderMenuItem, {
          href: item.url,
          "data-autoid": "".concat(stablePrefix, "--masthead-").concat(navType, "__l1-nav").concat(i, "-item").concat(j),
          key: item.title
        }, item.title));
      });
    });
  }

  return navItems;
}

MastheadL1.propTypes = {
  /**
   * The title (experimental).
   */
  title: PropTypes.string,

  /**
   * The optional title link (experimental)
   */
  titleLink: PropTypes.string,

  /**
   * Object containing masthead l1 navigation elements.
   */
  navigationL1: PropTypes.arrayOf(PropTypes.shape({
    hasMenu: PropTypes.bool,
    title: PropTypes.string,
    url: PropTypes.string,
    menuSections: PropTypes.arrayOf(PropTypes.shape({
      menuItems: PropTypes.arrayOf(PropTypes.shape({
        title: PropTypes.string,
        url: PropTypes.string
      }))
    }))
  }))
};
MastheadL1.defaultProps = {
  navigationL1: [],
  titleLink: null
};
export default MastheadL1;
