// @carbon/ibmdotcom-react/es/components/ContentSection/ContentSection.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import classNames from 'classnames';
import { CTA } from '../CTA';
import ddsSettings from '../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings';
import markdownToHtml from '@carbon/ibmdotcom-utilities/es/utilities/markdownToHtml/markdownToHtml';
import PropTypes from 'prop-types';
import React from 'react';
import settings from 'carbon-components/es/globals/js/settings';
var stablePrefix = ddsSettings.stablePrefix;
var prefix = settings.prefix;
/**
 * ContentSection Component, for use with cardSection.
 */

var ContentSection = function ContentSection(_ref) {
  var heading = _ref.heading,
      copy = _ref.copy,
      theme = _ref.theme,
      children = _ref.children,
      cta = _ref.cta,
      customClassName = _ref.customClassName,
      childrenCustomClassName = _ref.childrenCustomClassName,
      otherProps = _objectWithoutProperties(_ref, ["heading", "copy", "theme", "children", "cta", "customClassName", "childrenCustomClassName"]);

  return /*#__PURE__*/React.createElement("section", {
    className: classNames("".concat(prefix, "--content-section"), customClassName, _defineProperty({}, "".concat(prefix, "--content-section--").concat(theme), theme)),
    "data-autoid": otherProps.autoid ? otherProps.autoid : "".concat(stablePrefix, "--content-section")
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--content-section__grid")
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--content-section__row")
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--content-section__left")
  }, heading && /*#__PURE__*/React.createElement("h2", {
    className: "".concat(prefix, "--content-section__heading")
  }, heading), copy && /*#__PURE__*/React.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--content-section__copy"),
    className: "".concat(prefix, "--content-section__copy"),
    dangerouslySetInnerHTML: {
      __html: markdownToHtml(copy, {
        bold: false
      })
    }
  }), cta && /*#__PURE__*/React.createElement(CTA, {
    style: "text",
    type: cta.type,
    copy: cta.copy,
    href: cta.href,
    customClassName: "".concat(prefix, "--content-section__cta")
  })), /*#__PURE__*/React.createElement("div", {
    className: classNames("".concat(prefix, "--content-section__children"), childrenCustomClassName)
  }, children))));
};

ContentSection.propTypes = {
  /**
   * Heading text.
   */
  heading: PropTypes.string.isRequired,

  /**
   * Copy text.
   */
  copy: PropTypes.string,

  /**
   * Theme name.
   */
  theme: PropTypes.oneOf(['white', 'g10', 'g90', 'g100']),

  /**
   * Container for other components.
   */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),

  /**
   * CTA object.
   * See the [`<CTA>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-cta--default#props) for full usage details.
   */
  cta: PropTypes.shape({
    type: PropTypes.oneOfType([PropTypes.oneOf(['local', 'external', 'download', 'video']), PropTypes.arrayOf(PropTypes.oneOf(['local', 'external', 'download', 'video']))]),
    copy: PropTypes.string,
    href: PropTypes.string,
    customClassName: PropTypes.string
  }),

  /**
   * Optional class to be applied to the containing node.
   */
  customClassName: PropTypes.string,

  /**
   * Optional class to be applied to the child node.
   */
  childrenCustomClassName: PropTypes.string
};
export default ContentSection;
