// @carbon/ibmdotcom-react/lib/components/AudioPlayer/AudioPlayerThumbnail.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _FeatureFlags = require("../../internal/FeatureFlags");

var _services = require("@carbon/ibmdotcom-services/lib/services");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _settings = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var prefix = _settings.default.prefix;

var AudioPlayerThumbnail = function AudioPlayerThumbnail(_ref) {
  var audioId = _ref.audioId;
  var inputRef = (0, _react.useRef)(null);

  var thumbnailUrl = _services.KalturaPlayer.getThumbnailUrl({
    mediaId: audioId,
    width: '48',
    height: '48'
  });

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--audio-player__thumbnail-container"),
    ref: inputRef
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--audio-player__thumbnail"),
    style: {
      backgroundImage: "url('".concat(thumbnailUrl, "')")
    }
  })));
};

AudioPlayerThumbnail.propTypes = {
  /**
   * The Media ID from Kaltura media platform.
   */
  audioId: _propTypes.default.string.isRequired
};
AudioPlayerThumbnail.defaultProps = {};

var _default = !_FeatureFlags.DDS_AUDIO_PLAYER ? undefined : AudioPlayerThumbnail;

exports.default = _default;
