// @carbon/ibmdotcom-react/lib/components/AudioPlayer/AudioPlayerCaptionsMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _AudioPlayerDisabledButton = _interopRequireDefault(require("./AudioPlayerDisabledButton"));

var _ = _interopRequireDefault(require("@carbon/icons-react/lib/closed-caption--filled/20"));

var _FeatureFlags = require("../../internal/FeatureFlags");

var _OverflowMenu = _interopRequireDefault(require("../../internal/vendor/carbon-components-react/components/OverflowMenu"));

var _OverflowMenuItem = _interopRequireDefault(require("../../internal/vendor/carbon-components-react/components/OverflowMenuItem"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _settings = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var prefix = _settings.default.prefix;

var AudioPlayerCaptionsMenu = function AudioPlayerCaptionsMenu(_ref) {
  var kalturaDigitalPlayer = _ref.kalturaDigitalPlayer,
      availableCaptions = _ref.availableCaptions,
      audioCaption = _ref.audioCaption,
      setAudioCaption = _ref.setAudioCaption,
      setDisplayVolumeControl = _ref.setDisplayVolumeControl;

  var hasCaptions = function hasCaptions() {
    return Object.keys(availableCaptions).length > 0;
  };

  var handleAudioCaptions = function handleAudioCaptions() {
    var caption = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    setAudioCaption(caption);
  };

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, !kalturaDigitalPlayer || !hasCaptions() ? /*#__PURE__*/_react.default.createElement(_AudioPlayerDisabledButton.default, {
    icon: _.default
  }) : /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--audio-player__overflow-menu-container")
  }, /*#__PURE__*/_react.default.createElement(_OverflowMenu.default, {
    renderIcon: _.default,
    direction: "top",
    flipped: true,
    iconDescription: "Closed Captions",
    selectorPrimaryFocus: ".".concat(prefix, "--audio-player__button-for-closed-caption-").concat(audioCaption.toLowerCase()),
    onOpen: function onOpen() {
      return setDisplayVolumeControl(false);
    }
  }, /*#__PURE__*/_react.default.createElement(_OverflowMenuItem.default, {
    className: "".concat(prefix, "--audio-player__button-for-closed-caption-"),
    itemText: "Off",
    hasDivider: true,
    onClick: function onClick() {
      return handleAudioCaptions();
    },
    disabled: !kalturaDigitalPlayer || !hasCaptions()
  }), Object.keys(availableCaptions).map(function (captionLabel, captionIndex) {
    return /*#__PURE__*/_react.default.createElement(_OverflowMenuItem.default, {
      className: "".concat(prefix, "--audio-player__button-for-closed-caption-").concat(captionLabel.toLowerCase()),
      key: captionIndex,
      itemText: captionLabel,
      hasDivider: true,
      onClick: function onClick() {
        return handleAudioCaptions(captionLabel.toString());
      },
      disabled: !kalturaDigitalPlayer || !hasCaptions()
    });
  }))));
};

AudioPlayerCaptionsMenu.propTypes = {
  /**
   * The kaltura digital player (KDP) object
   * It starts as false and gets morphed into the html element
   *  of the target player id reference during the kaltura player
   *  embeding process as soon as the kaltura ready callback triggers
   */
  kalturaDigitalPlayer: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.bool]).isRequired,

  /**
   * An object containing all captions.
   * Each key of the object is the caption language label
   *  and contain an array of entries.
   * Each entry of this array is an object containing:
   *  start -> Number - The first second of the audio where
   *          the caption should be displayed
   *  end -> Number - The last second of the audio where
   *          the caption should be displayed
   *  content -> String - The html content of the caption
   */
  availableCaptions: _propTypes.default.object.isRequired,

  /**
   * The state getter for the current selected caption language
   */
  audioCaption: _propTypes.default.string.isRequired,

  /**
   * The state setter for the current selected caption language
   */
  setAudioCaption: _propTypes.default.func.isRequired,

  /**
   * The state setter that show/hide the custom volume menu
   */
  setDisplayVolumeControl: _propTypes.default.func.isRequired
};
AudioPlayerCaptionsMenu.defaultProps = {
  kalturaDigitalPlayer: false,
  availableCaptions: [],
  audioCaption: ''
};

var _default = !_FeatureFlags.DDS_AUDIO_PLAYER ? undefined : AudioPlayerCaptionsMenu;

exports.default = _default;
