// @carbon/ibmdotcom-react/lib/components/AudioPlayer/AudioPlayerPlaybackRateMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _AudioPlayerDisabledButton = _interopRequireDefault(require("./AudioPlayerDisabledButton"));

var _FeatureFlags = require("../../internal/FeatureFlags");

var _OverflowMenu = _interopRequireDefault(require("../../internal/vendor/carbon-components-react/components/OverflowMenu"));

var _OverflowMenuItem = _interopRequireDefault(require("../../internal/vendor/carbon-components-react/components/OverflowMenuItem"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _settings = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

var _ = _interopRequireDefault(require("@carbon/icons-react/lib/time/32"));

/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var prefix = _settings.default.prefix;

var AudioPlayerPlaybackRateMenu = function AudioPlayerPlaybackRateMenu(_ref) {
  var kalturaDigitalPlayer = _ref.kalturaDigitalPlayer,
      availablePlaybackRates = _ref.availablePlaybackRates,
      audioPlaybackRate = _ref.audioPlaybackRate,
      setAudioPlaybackRate = _ref.setAudioPlaybackRate,
      setDisplayVolumeControl = _ref.setDisplayVolumeControl;

  var handleAudioPlaybackRate = function handleAudioPlaybackRate(velocity) {
    if (kalturaDigitalPlayer) {
      setAudioPlaybackRate(velocity);
      kalturaDigitalPlayer.sendNotification('playbackRateChangeSpeed', velocity.toString());
    }
  };

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, !kalturaDigitalPlayer ? /*#__PURE__*/_react.default.createElement(_AudioPlayerDisabledButton.default, {
    icon: _.default
  }) : /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--audio-player__overflow-menu-container")
  }, /*#__PURE__*/_react.default.createElement(_OverflowMenu.default, {
    renderIcon: _.default,
    direction: "top",
    flipped: true,
    iconDescription: "Playback Rate Speed",
    selectorPrimaryFocus: ".".concat(prefix, "--audio-player__button-for-playback-rate-").concat(audioPlaybackRate.toString().split('.').join('-'), "x"),
    onOpen: function onOpen() {
      return setDisplayVolumeControl(false);
    }
  }, availablePlaybackRates.sort(function (a, b) {
    return a - b;
  }).map(function (playbackRateValue, playbackRateIndex) {
    return /*#__PURE__*/_react.default.createElement(_OverflowMenuItem.default, {
      className: "".concat(prefix, "--audio-player__button-for-playback-rate-").concat(playbackRateValue.toString().split('.').join('-'), "x"),
      key: playbackRateIndex,
      itemText: playbackRateValue.toString() + 'x',
      hasDivider: true,
      onClick: function onClick() {
        return handleAudioPlaybackRate(playbackRateValue);
      },
      disabled: !kalturaDigitalPlayer
    });
  }))));
};

AudioPlayerPlaybackRateMenu.propTypes = {
  /**
   * The kaltura digital player (KDP) object
   * It starts as false and gets morphed into the html element
   *  of the target player id reference during the kaltura player
   *  embeding process as soon as the kaltura ready callback triggers
   */
  kalturaDigitalPlayer: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.bool]).isRequired,

  /**
   * An array containing the available velocities/rates of the playback
   * Decimal - Ideally the decimals should be multiples of 0.25
   * Example & Default Values (inherited from parent):
   *  [1, 1.5, 2]
   */
  availablePlaybackRates: _propTypes.default.arrayOf(_propTypes.default.number).isRequired,

  /**
   * The state getter for the current playback rate
   */
  audioPlaybackRate: _propTypes.default.number.isRequired,

  /**
   * The state setter for the current playback rate
   */
  setAudioPlaybackRate: _propTypes.default.func.isRequired,

  /**
   * The state setter that show/hide the custom volume menu
   */
  setDisplayVolumeControl: _propTypes.default.func.isRequired
};
AudioPlayerPlaybackRateMenu.defaultProps = {
  kalturaDigitalPlayer: false
};

var _default = !_FeatureFlags.DDS_AUDIO_PLAYER ? undefined : AudioPlayerPlaybackRateMenu;

exports.default = _default;
