// @carbon/ibmdotcom-react/lib/components/AudioPlayer/AudioPlayerDisabledButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Button = _interopRequireDefault(require("../../internal/vendor/carbon-components-react/components/Button/Button"));

var _FeatureFlags = require("../../internal/FeatureFlags");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var AudioPlayerDisabledButton = function AudioPlayerDisabledButton(_ref) {
  var icon = _ref.icon;
  return /*#__PURE__*/_react.default.createElement(_Button.default, {
    renderIcon: icon,
    iconDescription: "Disabled",
    hasIconOnly: true,
    kind: "ghost",
    disabled: true
  });
};

AudioPlayerDisabledButton.propTypes = {
  /**
   * The icon that the button should render
   */
  icon: _propTypes.default.object
};

var _default = !_FeatureFlags.DDS_AUDIO_PLAYER ? undefined : AudioPlayerDisabledButton;

exports.default = _default;
