// @carbon/ibmdotcom-react/lib/components/ContentSection/ContentSection.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classnames = _interopRequireDefault(require("classnames"));

var _CTA = require("../CTA");

var _settings = _interopRequireDefault(require("../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings"));

var _markdownToHtml = _interopRequireDefault(require("@carbon/ibmdotcom-utilities/lib/utilities/markdownToHtml/markdownToHtml"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _settings2 = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var stablePrefix = _settings.default.stablePrefix;
var prefix = _settings2.default.prefix;
/**
 * ContentSection Component, for use with cardSection.
 */

var ContentSection = function ContentSection(_ref) {
  var heading = _ref.heading,
      copy = _ref.copy,
      theme = _ref.theme,
      children = _ref.children,
      cta = _ref.cta,
      customClassName = _ref.customClassName,
      childrenCustomClassName = _ref.childrenCustomClassName,
      otherProps = (0, _objectWithoutProperties2.default)(_ref, ["heading", "copy", "theme", "children", "cta", "customClassName", "childrenCustomClassName"]);
  return /*#__PURE__*/_react.default.createElement("section", {
    className: (0, _classnames.default)("".concat(prefix, "--content-section"), customClassName, (0, _defineProperty2.default)({}, "".concat(prefix, "--content-section--").concat(theme), theme)),
    "data-autoid": otherProps.autoid ? otherProps.autoid : "".concat(stablePrefix, "--content-section")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--content-section__grid")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--content-section__row")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--content-section__left")
  }, heading && /*#__PURE__*/_react.default.createElement("h2", {
    className: "".concat(prefix, "--content-section__heading")
  }, heading), copy && /*#__PURE__*/_react.default.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--content-section__copy"),
    className: "".concat(prefix, "--content-section__copy"),
    dangerouslySetInnerHTML: {
      __html: (0, _markdownToHtml.default)(copy, {
        bold: false
      })
    }
  }), cta && /*#__PURE__*/_react.default.createElement(_CTA.CTA, {
    style: "text",
    type: cta.type,
    copy: cta.copy,
    href: cta.href,
    customClassName: "".concat(prefix, "--content-section__cta")
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)("".concat(prefix, "--content-section__children"), childrenCustomClassName)
  }, children))));
};

ContentSection.propTypes = {
  /**
   * Heading text.
   */
  heading: _propTypes.default.string.isRequired,

  /**
   * Copy text.
   */
  copy: _propTypes.default.string,

  /**
   * Theme name.
   */
  theme: _propTypes.default.oneOf(['white', 'g10', 'g90', 'g100']),

  /**
   * Container for other components.
   */
  children: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.node), _propTypes.default.node]),

  /**
   * CTA object.
   * See the [`<CTA>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-cta--default#props) for full usage details.
   */
  cta: _propTypes.default.shape({
    type: _propTypes.default.oneOfType([_propTypes.default.oneOf(['local', 'external', 'download', 'video']), _propTypes.default.arrayOf(_propTypes.default.oneOf(['local', 'external', 'download', 'video']))]),
    copy: _propTypes.default.string,
    href: _propTypes.default.string,
    customClassName: _propTypes.default.string
  }),

  /**
   * Optional class to be applied to the containing node.
   */
  customClassName: _propTypes.default.string,

  /**
   * Optional class to be applied to the child node.
   */
  childrenCustomClassName: _propTypes.default.string
};
var _default = ContentSection;
exports.default = _default;
