// @carbon/ibmdotcom-react/es/components/Footer/LegalNav.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import cx from 'classnames';
import ddsSettings from '../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings';
import Link from '../../internal/vendor/carbon-components-react/components/Link/Link';
import PropTypes from 'prop-types';
import React from 'react';
import settings from 'carbon-components/es/globals/js/settings';
var stablePrefix = ddsSettings.stablePrefix;
var prefix = settings.prefix;
/**
 * Footer legal nav component.
 */

var LegalNav = function LegalNav(_ref) {
  var links = _ref.links,
      type = _ref.type,
      button = _ref.button,
      adjunctLinks = _ref.adjunctLinks;

  if (!(links === null || links === void 0 ? void 0 : links.length)) {
    return null;
  }

  var listStyle = cx(_defineProperty({}, "".concat(prefix, "--legal-nav__micro"), type === 'micro'));
  return /*#__PURE__*/React.createElement("aside", {
    "data-autoid": "".concat(stablePrefix, "--footer-legal-nav"),
    className: "".concat(prefix, "--legal-nav__container"),
    "aria-label": "Footer legal links container"
  }, /*#__PURE__*/React.createElement("nav", {
    className: "".concat(prefix, "--legal-nav"),
    "aria-label": "Footer legal link list"
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--legal-nav__list ").concat(listStyle)
  }, /*#__PURE__*/React.createElement("ul", {
    className: "".concat(prefix, "--legal-nav__holder")
  }, links && links.map(function (_ref2, index) {
    var title = _ref2.title,
        titleEnglish = _ref2.titleEnglish,
        url = _ref2.url;

    if (!title || !url) {
      return null;
    }

    var dataTitle = titleEnglish ? titleEnglish.replace(/[^-a-zA-Z0-9_ ]/g, '').replace(/ +/g, '-').toLowerCase() : null;
    return /*#__PURE__*/React.createElement("li", {
      className: "".concat(prefix, "--legal-nav__list-item"),
      key: index
    }, /*#__PURE__*/React.createElement(Link, {
      "data-autoid": "".concat(stablePrefix, "--footer-legal-nav__link-").concat(dataTitle),
      className: "".concat(prefix, "--footer__link"),
      href: url
    }, title));
  }), /*#__PURE__*/React.createElement("li", {
    className: "".concat(prefix, "--legal-nav__list-item"),
    "data-autoid": "".concat(stablePrefix, "--privacy-cp")
  })), button), adjunctLinks && /*#__PURE__*/React.createElement("ul", {
    className: "".concat(prefix, "--adjunct-links__holder")
  }, adjunctLinks.map(function (_ref3, index) {
    var title = _ref3.title,
        titleEnglish = _ref3.titleEnglish,
        url = _ref3.url;

    if (!title || !url) {
      return null;
    }

    var dataTitle = titleEnglish ? titleEnglish.replace(/[^-a-zA-Z0-9_ ]/g, '').replace(/ +/g, '-').toLowerCase() : null;
    return /*#__PURE__*/React.createElement("li", {
      className: "".concat(prefix, "--legal-nav__list-item"),
      key: index
    }, /*#__PURE__*/React.createElement(Link, {
      "data-autoid": "".concat(stablePrefix, "--footer-legal-nav__link-").concat(dataTitle),
      className: "".concat(prefix, "--footer__link"),
      href: url
    }, title));
  }))));
};

LegalNav.propTypes = {
  /**
   * A list of links to be rendered.
   */
  links: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string,
    url: PropTypes.string
  })),

  /**
   * List of adjunct links to be rendered
   */
  adjunctLinks: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string,
    url: PropTypes.string
  })),

  /**
   * Footer type
   */
  type: PropTypes.string,

  /**
   * The locale/language selector button.
   * Renders only in micro version
   */
  button: PropTypes.object
};
LegalNav.defaultProps = {
  links: null
};
export default LegalNav;
