// @carbon/ibmdotcom-react/es/components/AudioPlayer/AudioPlayerThumbnail.js
/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React, { useRef } from 'react';
import { DDS_AUDIO_PLAYER } from '../../internal/FeatureFlags';
import { KalturaPlayer as KalturaPlayerAPI } from '@carbon/ibmdotcom-services/es/services';
import PropTypes from 'prop-types';
import settings from 'carbon-components/es/globals/js/settings';
var prefix = settings.prefix;

var AudioPlayerThumbnail = function AudioPlayerThumbnail(_ref) {
  var audioId = _ref.audioId;
  var inputRef = useRef(null);
  var thumbnailUrl = KalturaPlayerAPI.getThumbnailUrl({
    mediaId: audioId,
    width: '48',
    height: '48'
  });
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--audio-player__thumbnail-container"),
    ref: inputRef
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--audio-player__thumbnail"),
    style: {
      backgroundImage: "url('".concat(thumbnailUrl, "')")
    }
  })));
};

AudioPlayerThumbnail.propTypes = {
  /**
   * The Media ID from Kaltura media platform.
   */
  audioId: PropTypes.string.isRequired
};
AudioPlayerThumbnail.defaultProps = {};
export default !DDS_AUDIO_PLAYER ? undefined : AudioPlayerThumbnail;
