// @carbon/ibmdotcom-react/lib/components/Card/Card.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cardPropTypes = exports.Card = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _CTALogic = _interopRequireDefault(require("../CTA/CTALogic"));

var _settings = _interopRequireDefault(require("../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings"));

var _Image = require("../Image");

var _Link = _interopRequireDefault(require("../../internal/vendor/carbon-components-react/components/Link/Link"));

var _markdownToHtml = _interopRequireDefault(require("@carbon/ibmdotcom-utilities/lib/utilities/markdownToHtml/markdownToHtml"));

var _on = _interopRequireDefault(require("carbon-components/umd/globals/js/misc/on"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _settings2 = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

var _Tile = require("../../internal/vendor/carbon-components-react/components/Tile/Tile");

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var stablePrefix = _settings.default.stablePrefix;
var prefix = _settings2.default.prefix;
/**
 * Card Link Component.
 */

var Card = function Card(_ref) {
  var _classNames;

  var cardStatic = _ref.cardStatic,
      light = _ref.light,
      inverse = _ref.inverse,
      border = _ref.border,
      image = _ref.image,
      eyebrow = _ref.eyebrow,
      heading = _ref.heading,
      customClassName = _ref.customClassName,
      copy = _ref.copy,
      cta = _ref.cta,
      pictogram = _ref.pictogram,
      onClick = _ref.onClick,
      props = (0, _objectWithoutProperties2.default)(_ref, ["cardStatic", "light", "inverse", "border", "image", "eyebrow", "heading", "customClassName", "copy", "cta", "pictogram", "onClick"]);
  var refWrapper = (0, _react.useRef)(null);
  var handleClick = (0, _react.useCallback)(function (e) {
    if (!onClick || onClick(e) !== false) {
      (cta === null || cta === void 0 ? void 0 : cta.type) === 'jump' ? _CTALogic.default.jump(e, cta.type) : false;
    }
  }, [cta, onClick]);
  (0, _react.useEffect)(function () {
    var hClick;
    var wrapperNode = refWrapper.current;
    var tileNode = wrapperNode.closest('.bx--tile');

    if (tileNode) {
      // Manually attach an event listener given `onClick()` of Carbon `<Tile>` runs after `<Tile>` changes its state
      hClick = (0, _on.default)(tileNode, 'click', handleClick);
    }

    return function () {
      if (hClick) {
        hClick = hClick.release();
      }
    };
  }, [handleClick]);
  return /*#__PURE__*/_react.default.createElement(_Tile.Tile, (0, _extends2.default)({
    "data-autoid": "".concat(stablePrefix, "--card"),
    className: (0, _classnames.default)("".concat(prefix, "--card"), (_classNames = {}, (0, _defineProperty2.default)(_classNames, "".concat(prefix, "--card--static"), cardStatic), (0, _defineProperty2.default)(_classNames, "".concat(prefix, "--card--light"), light), (0, _defineProperty2.default)(_classNames, "".concat(prefix, "--card--inverse"), inverse), (0, _defineProperty2.default)(_classNames, "".concat(prefix, "--card__CTA--disabled"), props.disabled), (0, _defineProperty2.default)(_classNames, "".concat(prefix, "--card--border"), border), _classNames), customClassName)
  }, props), image && /*#__PURE__*/_react.default.createElement(_Image.Image, (0, _extends2.default)({}, image, {
    classname: "".concat(prefix, "--card__img")
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--card__wrapper"),
    ref: refWrapper
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--card__content")
  }, eyebrow && /*#__PURE__*/_react.default.createElement("p", {
    className: "".concat(prefix, "--card__eyebrow")
  }, eyebrow), heading && /*#__PURE__*/_react.default.createElement("h3", {
    className: "".concat(prefix, "--card__heading")
  }, heading), optionalContent(copy), renderFooter(cta, copy, props.disabled, heading, pictogram))));
};
/**
 * Card Link optional content
 *
 * @param {string} copy paragraph of text
 * @returns {object} JSX object
 */


exports.Card = Card;

function optionalContent(copy) {
  return !copy ? null : /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--card__copy"),
    dangerouslySetInnerHTML: {
      __html: (0, _markdownToHtml.default)(copy, {
        bold: false
      })
    }
  });
}
/**
 * Render footer with icon
 *
 * @param {object} cta cta object
 * @returns {object} JSX object
 */


function renderFooter(cta, copy, disabled, heading, pictogram) {
  var _classNames2, _cta$icon;

  var CardFooter = disabled ? 'p' : _Link.default;
  return cta && /*#__PURE__*/_react.default.createElement(CardFooter, {
    className: (0, _classnames.default)("".concat(prefix, "--card__footer"), (_classNames2 = {}, (0, _defineProperty2.default)(_classNames2, "".concat(prefix, "--card__footer__icon-left"), (cta === null || cta === void 0 ? void 0 : cta.iconPlacement) === 'left'), (0, _defineProperty2.default)(_classNames2, "".concat(prefix, "--card__footer__copy"), cta === null || cta === void 0 ? void 0 : cta.copy), _classNames2)),
    href: cta === null || cta === void 0 ? void 0 : cta.href,
    target: _CTALogic.default.external(cta === null || cta === void 0 ? void 0 : cta.type),
    "aria-label": ((cta === null || cta === void 0 ? void 0 : cta.copy) ? '' : heading ? heading : copy) + _CTALogic.default.getDefaultLabel(cta === null || cta === void 0 ? void 0 : cta.type),
    onClick: cta === null || cta === void 0 ? void 0 : cta.handleClick
  }, (cta === null || cta === void 0 ? void 0 : cta.copy) && !pictogram && /*#__PURE__*/_react.default.createElement("span", {
    className: "".concat(prefix, "--card__cta__copy")
  }, cta === null || cta === void 0 ? void 0 : cta.copy), (cta === null || cta === void 0 ? void 0 : (_cta$icon = cta.icon) === null || _cta$icon === void 0 ? void 0 : _cta$icon.src) && !pictogram && /*#__PURE__*/_react.default.createElement(cta.icon.src, (0, _extends2.default)({
    className: "".concat(prefix, "--card__cta")
  }, cta === null || cta === void 0 ? void 0 : cta.icon)), pictogram && pictogram);
}

var cardPropTypes = {
  /**
   * Concise yet descriptive string of text describing the linked resource.
   */
  heading: _propTypes.default.string,

  /**
   * Eyebrow text to be passed as a property to the Card component.
   * Style for Card component has been set to "text".
   */
  eyebrow: _propTypes.default.string,

  /**
   * Paragraph of text that further describing the resource with added detail.
   */
  copy: _propTypes.default.string,

  /**
   * Disable card link
   */
  disabled: _propTypes.default.bool,

  /**
   * Pictogram located at the bottom left side of the Card. This prop disables the CTA.copy and CTA.icon (experimental)
   */
  pictogram: _propTypes.default.node,

  /**
   * CTA options. Has the following structure in summary:
   *
   * | Name            | Data Type | Description                                                                                                                      |
   * | --------------- | --------- | -------------------------------------------------------------------------------------------------------------------------------- |
   * | `href`          | String    | Valid URL for a the location of an internal or external resource                                                                 |
   * | `icon`          | String    | Provide an optional icon to the footer from [Carbon's icon library](https://www.carbondesignsystem.com/guidelines/icons/library) |
   * | `iconPlacement` | String    | Option to place icon left or right of cta text                                                                                   |
   * | `copy`          | String    | Optional text for CTA                                                                                                            |
   * | `type`          | String    | type of CTA (local or external) when Card type is static                                                                         |
   *
   * See the [`<CTA>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-cta--default#props) for full usage details.
   */
  cta: _propTypes.default.shape({
    copy: _propTypes.default.string,
    href: _propTypes.default.string,
    type: _propTypes.default.oneOf(['jump', 'local', 'external', 'download']),
    icon: _propTypes.default.shape({
      src: _propTypes.default.elementType
    }),
    iconPlacement: _propTypes.default.oneOf(['left', 'right'])
  }),

  /**
   * Contains source and alt text properties.
   * See [`<Image>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-image--default#props) for full usage details.
   */
  image: _propTypes.default.shape({
    classname: _propTypes.default.string,
    sources: _propTypes.default.arrayOf(_propTypes.default.shape({
      src: _propTypes.default.string,
      breakpoint: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number])
    })),
    defaultSrc: _propTypes.default.string.isRequired,
    alt: _propTypes.default.string.isRequired,
    longDescription: _propTypes.default.string
  }),

  /**
   * `true` to set a 1px solid border around Card.
   */
  border: _propTypes.default.bool,

  /**
   * `true` to set the Card static variation.
   */
  cardStatic: _propTypes.default.bool,

  /**
   * `true` to set the high contrast for Card.
   */
  inverse: _propTypes.default.bool,

  /**
   * `true` to set the light theme for Card.
   */
  light: _propTypes.default.bool,

  /**
   * Classname to be assigned to the Card component.
   */
  customClassName: _propTypes.default.string,

  /**
   * A handler for `click` event on the card.
   */
  onClick: _propTypes.default.func
};
exports.cardPropTypes = cardPropTypes;
Card.propTypes = cardPropTypes;
