// @carbon/ibmdotcom-react/lib/components/Masthead/MastheadMegaMenu/CategoryLink.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _ = _interopRequireDefault(require("@carbon/icons-react/lib/arrow--right/16"));

var _Link = _interopRequireDefault(require("../../../internal/vendor/carbon-components-react/components/Link/Link"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _settings = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var prefix = _settings.default.prefix;
/**
 * Category sublink
 */

var CategoryLink = function CategoryLink(_ref) {
  var href = _ref.href,
      title = _ref.title,
      highlighted = _ref.highlighted,
      rest = (0, _objectWithoutProperties2.default)(_ref, ["href", "title", "highlighted"]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, highlighted ? /*#__PURE__*/_react.default.createElement(_Link.default, {
    className: "".concat(prefix, "--masthead__megamenu__category-sublink--highlighted"),
    href: href,
    renderIcon: _.default
  }, /*#__PURE__*/_react.default.createElement("span", null, title)) : /*#__PURE__*/_react.default.createElement("a", {
    tabIndex: 0,
    href: href,
    className: "".concat(prefix, "--masthead__megamenu__category-sublink"),
    "data-autoid": "".concat(rest.autoid, "-item").concat(rest.index)
  }, title));
};

CategoryLink.propTypes = {
  /**
   * Category sublink href
   */
  href: _propTypes.default.string.isRequired,

  /**
   * Category sublink text
   */
  title: _propTypes.default.string.isRequired,

  /**
   * Determines whether to render regular link style or link with icon
   */
  highlighted: _propTypes.default.bool
};
var _default = CategoryLink;
exports.default = _default;
