// @carbon/ibmdotcom-react/lib/components/Masthead/MastheadMegaMenu/MegaMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _CategoryGroup = _interopRequireDefault(require("./CategoryGroup"));

var _CategoryLink = _interopRequireDefault(require("./CategoryLink"));

var _LeftNavigation = _interopRequireDefault(require("./LeftNavigation"));

var _NavigationGroup = _interopRequireDefault(require("./NavigationGroup"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _RightNavigation = _interopRequireDefault(require("./RightNavigation"));

var _settings = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var prefix = _settings.default.prefix;
/**
 * Masthead megamenu component.
 */

var MegaMenu = _react.default.forwardRef(function MegaMenu(props, ref) {
  var _data$menuSections$, _viewAllLink, _viewAllLink2;

  var data = props.data,
      rest = (0, _objectWithoutProperties2.default)(props, ["data"]);
  var highlightedItems = [];
  var viewAllLink;
  var menuItems = [];
  data.menuSections[0].menuItems.forEach(function (item) {
    if (item.highlighted) return highlightedItems.push(item);
    if (item.megaPanelViewAll) return viewAllLink = item;else return menuItems.push(item);
  });
  var hasHighlights = highlightedItems.length !== 0;
  var heading = (_data$menuSections$ = data.menuSections[0]) === null || _data$menuSections$ === void 0 ? void 0 : _data$menuSections$.heading;
  return /*#__PURE__*/_react.default.createElement(_NavigationGroup.default, {
    ref: ref
  }, hasHighlights && /*#__PURE__*/_react.default.createElement(_LeftNavigation.default, null, heading && /*#__PURE__*/_react.default.createElement("p", {
    className: "".concat(prefix, "--masthead__megamenu__copy")
  }, heading), highlightedItems.map(function (item, i) {
    var _item$megapanelConten, _item$megapanelConten2, _item$megapanelConten3, _item$megapanelConten4;

    return /*#__PURE__*/_react.default.createElement(_CategoryGroup.default, {
      autoid: rest.autoid,
      key: i,
      href: item.url,
      title: item.title
    }, ((_item$megapanelConten = item.megapanelContent) === null || _item$megapanelConten === void 0 ? void 0 : _item$megapanelConten.description) && /*#__PURE__*/_react.default.createElement("p", {
      className: "".concat(prefix, "--masthead__megamenu__copy")
    }, (_item$megapanelConten2 = item.megapanelContent) === null || _item$megapanelConten2 === void 0 ? void 0 : _item$megapanelConten2.description), (_item$megapanelConten3 = item.megapanelContent) === null || _item$megapanelConten3 === void 0 ? void 0 : (_item$megapanelConten4 = _item$megapanelConten3.quickLinks) === null || _item$megapanelConten4 === void 0 ? void 0 : _item$megapanelConten4.links.map(function (_ref, key) {
      var title = _ref.title,
          url = _ref.url,
          highlightedLink = _ref.highlightedLink;
      return /*#__PURE__*/_react.default.createElement(_CategoryLink.default, {
        href: url,
        title: title,
        highlighted: highlightedLink,
        autoid: "".concat(rest.autoid, "-list").concat(i),
        key: key
      });
    }));
  })), /*#__PURE__*/_react.default.createElement(_RightNavigation.default, {
    hasHighlights: hasHighlights,
    viewAllLinkHref: (_viewAllLink = viewAllLink) === null || _viewAllLink === void 0 ? void 0 : _viewAllLink.url,
    viewAllLinkTitle: (_viewAllLink2 = viewAllLink) === null || _viewAllLink2 === void 0 ? void 0 : _viewAllLink2.title,
    autoid: rest.autoid
  }, menuItems.map(function (item, i) {
    var _item$megapanelConten5, _item$megapanelConten6;

    return /*#__PURE__*/_react.default.createElement(_CategoryGroup.default, {
      key: i,
      autoid: rest.autoid,
      index: i + highlightedItems.length,
      href: item.url,
      title: item.title
    }, (_item$megapanelConten5 = item.megapanelContent) === null || _item$megapanelConten5 === void 0 ? void 0 : (_item$megapanelConten6 = _item$megapanelConten5.quickLinks) === null || _item$megapanelConten6 === void 0 ? void 0 : _item$megapanelConten6.links.map(function (_ref2, key) {
      var title = _ref2.title,
          url = _ref2.url;
      return /*#__PURE__*/_react.default.createElement(_CategoryLink.default, {
        key: key,
        href: url,
        title: title,
        autoid: "".concat(rest.autoid, "-list").concat(i + highlightedItems.length),
        index: key
      });
    }));
  })));
});

MegaMenu.propTypes = {
  /**
   * Object containing megamenu nav data
   */
  data: _propTypes.default.shape({
    hasMenupanel: _propTypes.default.bool,
    title: _propTypes.default.string,
    url: _propTypes.default.string,
    menuSections: _propTypes.default.arrayOf(_propTypes.default.shape({
      heading: _propTypes.default.string,
      menuItems: _propTypes.default.arrayOf(_propTypes.default.shape({
        highlighted: _propTypes.default.bool,
        title: _propTypes.default.string,
        url: _propTypes.default.string,
        megapanelContent: _propTypes.default.shape({
          quickLinks: _propTypes.default.shape({
            links: _propTypes.default.arrayOf(_propTypes.default.shape({
              title: _propTypes.default.string,
              url: _propTypes.default.string
            }))
          })
        })
      }))
    }))
  })
};
var _default = MegaMenu;
exports.default = _default;
