// @carbon/ibmdotcom-react/lib/components/Footer/LegalNav.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classnames = _interopRequireDefault(require("classnames"));

var _settings = _interopRequireDefault(require("../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings"));

var _Link = _interopRequireDefault(require("../../internal/vendor/carbon-components-react/components/Link/Link"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _settings2 = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var stablePrefix = _settings.default.stablePrefix;
var prefix = _settings2.default.prefix;
/**
 * Footer legal nav component.
 */

var LegalNav = function LegalNav(_ref) {
  var links = _ref.links,
      type = _ref.type,
      button = _ref.button,
      adjunctLinks = _ref.adjunctLinks;

  if (!(links === null || links === void 0 ? void 0 : links.length)) {
    return null;
  }

  var listStyle = (0, _classnames.default)((0, _defineProperty2.default)({}, "".concat(prefix, "--legal-nav__micro"), type === 'micro'));
  return /*#__PURE__*/_react.default.createElement("aside", {
    "data-autoid": "".concat(stablePrefix, "--footer-legal-nav"),
    className: "".concat(prefix, "--legal-nav__container"),
    "aria-label": "Footer legal links container"
  }, /*#__PURE__*/_react.default.createElement("nav", {
    className: "".concat(prefix, "--legal-nav"),
    "aria-label": "Footer legal link list"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--legal-nav__list ").concat(listStyle)
  }, /*#__PURE__*/_react.default.createElement("ul", {
    className: "".concat(prefix, "--legal-nav__holder")
  }, links && links.map(function (_ref2, index) {
    var title = _ref2.title,
        titleEnglish = _ref2.titleEnglish,
        url = _ref2.url;

    if (!title || !url) {
      return null;
    }

    var dataTitle = titleEnglish ? titleEnglish.replace(/[^-a-zA-Z0-9_ ]/g, '').replace(/ +/g, '-').toLowerCase() : null;
    return /*#__PURE__*/_react.default.createElement("li", {
      className: "".concat(prefix, "--legal-nav__list-item"),
      key: index
    }, /*#__PURE__*/_react.default.createElement(_Link.default, {
      "data-autoid": "".concat(stablePrefix, "--footer-legal-nav__link-").concat(dataTitle),
      className: "".concat(prefix, "--footer__link"),
      href: url
    }, title));
  }), /*#__PURE__*/_react.default.createElement("li", {
    className: "".concat(prefix, "--legal-nav__list-item"),
    "data-autoid": "".concat(stablePrefix, "--privacy-cp")
  })), button), adjunctLinks && /*#__PURE__*/_react.default.createElement("ul", {
    className: "".concat(prefix, "--adjunct-links__holder")
  }, adjunctLinks.map(function (_ref3, index) {
    var title = _ref3.title,
        titleEnglish = _ref3.titleEnglish,
        url = _ref3.url;

    if (!title || !url) {
      return null;
    }

    var dataTitle = titleEnglish ? titleEnglish.replace(/[^-a-zA-Z0-9_ ]/g, '').replace(/ +/g, '-').toLowerCase() : null;
    return /*#__PURE__*/_react.default.createElement("li", {
      className: "".concat(prefix, "--legal-nav__list-item"),
      key: index
    }, /*#__PURE__*/_react.default.createElement(_Link.default, {
      "data-autoid": "".concat(stablePrefix, "--footer-legal-nav__link-").concat(dataTitle),
      className: "".concat(prefix, "--footer__link"),
      href: url
    }, title));
  }))));
};

LegalNav.propTypes = {
  /**
   * A list of links to be rendered.
   */
  links: _propTypes.default.arrayOf(_propTypes.default.shape({
    title: _propTypes.default.string,
    url: _propTypes.default.string
  })),

  /**
   * List of adjunct links to be rendered
   */
  adjunctLinks: _propTypes.default.arrayOf(_propTypes.default.shape({
    title: _propTypes.default.string,
    url: _propTypes.default.string
  })),

  /**
   * Footer type
   */
  type: _propTypes.default.string,

  /**
   * The locale/language selector button.
   * Renders only in micro version
   */
  button: _propTypes.default.object
};
LegalNav.defaultProps = {
  links: null
};
var _default = LegalNav;
exports.default = _default;
