// carbon-components-react/es/components/Tooltip/next/Tooltip.stories.js
/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import './story.scss';
import { Checkbox16 } from '@carbon/icons-react';
import React from 'react';
import { Tooltip } from '../next';
import { DefinitionTooltip } from './DefinitionTooltip.js';
export default {
  title: 'Experimental/unstable_Tooltip',
  component: Tooltip,
  parameters: {
    controls: {
      hideNoControlsWarning: true
    },
    layout: 'centered'
  },
  argTypes: {
    children: {
      table: {
        disable: true
      }
    },
    className: {
      table: {
        disable: true
      }
    }
  }
};
export var Default = function Default() {
  var label = 'Occassionally, services are updated in a specified time window to ensure no down time for customers.';
  return /*#__PURE__*/React.createElement(Tooltip, {
    align: "bottom",
    defaultOpen: true,
    label: label
  }, /*#__PURE__*/React.createElement("button", {
    className: "demo-tooltip-trigger",
    type: "button"
  }, /*#__PURE__*/React.createElement(Checkbox16, null)));
};
export var Duration = function Duration() {
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Tooltip, {
    label: "Label one",
    enterDelayMs: 500,
    leaveDelayMs: 500
  }, /*#__PURE__*/React.createElement("button", {
    className: "demo-tooltip-trigger",
    type: "button"
  }, /*#__PURE__*/React.createElement(Checkbox16, null))), /*#__PURE__*/React.createElement(Tooltip, {
    label: "Label two",
    enterDelayMs: 500,
    leaveDelayMs: 500
  }, /*#__PURE__*/React.createElement("button", {
    className: "demo-tooltip-trigger",
    type: "button"
  }, /*#__PURE__*/React.createElement(Checkbox16, null))), /*#__PURE__*/React.createElement(Tooltip, {
    label: "Label three",
    enterDelayMs: 500,
    leaveDelayMs: 500
  }, /*#__PURE__*/React.createElement("button", {
    className: "demo-tooltip-trigger",
    type: "button"
  }, /*#__PURE__*/React.createElement(Checkbox16, null))));
};
export var Definition = function Definition() {
  var definition = 'Uniform Resource Locator; the address of a resource (such as a document or website) on the Internet.';
  return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("p", null, "Custom domains direct requests for your apps in this Cloud Foundry organization to a", ' ', /*#__PURE__*/React.createElement(DefinitionTooltip, {
    definition: definition
  }, "URL"), " that you own. A custom domain can be a shared domain, a shared subdomain, or a shared domain and host."));
};

var PlaygroundStory = function PlaygroundStory(props) {
  var align = props.align,
      defaultOpen = props.defaultOpen,
      description = props.description,
      enterDelayMs = props.enterDelayMs,
      label = props.label,
      leaveDelayMs = props.leaveDelayMs;
  return /*#__PURE__*/React.createElement(Tooltip, {
    align: align,
    defaultOpen: defaultOpen,
    description: description,
    enterDelayMs: enterDelayMs,
    label: label,
    leaveDelayMs: leaveDelayMs
  }, /*#__PURE__*/React.createElement("button", {
    className: "demo-tooltip-trigger",
    type: "button"
  }, /*#__PURE__*/React.createElement(Checkbox16, null)));
};

export var Playground = PlaygroundStory.bind({});
Playground.argTypes = {
  align: {
    defaultValue: 'bottom',
    options: ['top', 'top-left', 'top-right', 'bottom', 'bottom-left', 'bottom-right', 'left', 'left-bottom', 'left-top', 'right', 'right-bottom', 'right-top'],
    control: {
      type: 'select'
    }
  },
  defaultOpen: {
    defaultValue: true
  },
  label: {
    control: {
      type: 'text'
    },
    defaultValue: 'Custom label'
  },
  description: {
    control: {
      type: 'text'
    }
  }
};
