// @carbon/ibmdotcom-react/es/components/Masthead/MastheadMegaMenu/CategoryLink.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";

/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import ArrowRight16 from '@carbon/icons-react/es/arrow--right/16';
import Link from '../../../internal/vendor/carbon-components-react/components/Link/Link';
import PropTypes from 'prop-types';
import React from 'react';
import settings from 'carbon-components/es/globals/js/settings';
var prefix = settings.prefix;
/**
 * Category sublink
 */

var CategoryLink = function CategoryLink(_ref) {
  var href = _ref.href,
      title = _ref.title,
      highlighted = _ref.highlighted,
      rest = _objectWithoutProperties(_ref, ["href", "title", "highlighted"]);

  return /*#__PURE__*/React.createElement(React.Fragment, null, highlighted ? /*#__PURE__*/React.createElement(Link, {
    className: "".concat(prefix, "--masthead__megamenu__category-sublink--highlighted"),
    href: href,
    renderIcon: ArrowRight16
  }, /*#__PURE__*/React.createElement("span", null, title)) : /*#__PURE__*/React.createElement("a", {
    tabIndex: 0,
    href: href,
    className: "".concat(prefix, "--masthead__megamenu__category-sublink"),
    "data-autoid": "".concat(rest.autoid, "-item").concat(rest.index)
  }, title));
};

CategoryLink.propTypes = {
  /**
   * Category sublink href
   */
  href: PropTypes.string.isRequired,

  /**
   * Category sublink text
   */
  title: PropTypes.string.isRequired,

  /**
   * Determines whether to render regular link style or link with icon
   */
  highlighted: PropTypes.bool
};
export default CategoryLink;
