// carbon-components-react/es/components/OverflowMenu/next/OverflowMenu.stories.js
/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import OverflowMenu from '../OverflowMenu';
import OverflowMenuItem from '../../OverflowMenuItem';
import mdx from '../OverflowMenu.mdx';
import { Filter16 } from '@carbon/icons-react';
export default {
  title: 'Components/OverflowMenu',
  component: OverflowMenu,
  subcomponents: {
    OverflowMenuItem: OverflowMenuItem
  },
  parameters: {
    docs: {
      page: mdx
    }
  }
};
export var Default = function Default() {
  return /*#__PURE__*/React.createElement(OverflowMenu, null, /*#__PURE__*/React.createElement(OverflowMenuItem, {
    itemText: "Stop app"
  }), /*#__PURE__*/React.createElement(OverflowMenuItem, {
    itemText: "Restart app"
  }), /*#__PURE__*/React.createElement(OverflowMenuItem, {
    itemText: "Rename app"
  }), /*#__PURE__*/React.createElement(OverflowMenuItem, {
    itemText: "Edit routes and access",
    requireTitle: true
  }), /*#__PURE__*/React.createElement(OverflowMenuItem, {
    hasDivider: true,
    isDelete: true,
    itemText: "Delete app"
  }));
};
Default.story = {
  name: 'OverflowMenu'
};
export var RenderCustomIcon = function RenderCustomIcon() {
  return /*#__PURE__*/React.createElement(OverflowMenu, {
    renderIcon: Filter16
  }, /*#__PURE__*/React.createElement(OverflowMenuItem, {
    itemText: "Filter A"
  }), /*#__PURE__*/React.createElement(OverflowMenuItem, {
    itemText: "Filter B"
  }));
};
