// @carbon/ibmdotcom-react/lib/components/Quote/Quote.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _ = _interopRequireDefault(require("@carbon/icons-react/lib/arrow--right/20"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _settings = _interopRequireDefault(require("../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings"));

var _decodeString = _interopRequireDefault(require("../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/decodeString/decodeString"));

var _HorizontalRule = require("../HorizontalRule");

var _2 = _interopRequireDefault(require("@carbon/icons-react/lib/launch/20"));

var _LinkWithIcon = require("../LinkWithIcon");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _settings2 = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var stablePrefix = _settings.default.stablePrefix;
var prefix = _settings2.default.prefix;
/**
 * Quote Pattern
 *
 * @param {object} props props Object
 * @param {string} props.markType type of the quote marks
 * @param {string} props.copy quote copy
 * @param {object} props.source source object
 * @param {string} props.source.heading name heading for quote source
 * @param {string} props.source.copy title copy for quote source
 * @param {object} props.cta cta props object
 * @param {string} props.cta.copy cta copy
 * @param {string} props.cta.type type 'local' or 'external'
 * @param {string} props.cta.href cta href
 * @param {boolean} props.inverse toggles inverse theme
 * @returns {*} Quote Pattern
 */

var Quote = function Quote(_ref) {
  var _ref$markType = _ref.markType,
      markType = _ref$markType === void 0 ? 'doubleCurved' : _ref$markType,
      copy = _ref.copy,
      source = _ref.source,
      cta = _ref.cta,
      inverse = _ref.inverse;

  var _useState = (0, _react.useState)(''),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      decodedCopy = _useState2[0],
      setDecodedCopy = _useState2[1];

  (0, _react.useEffect)(function () {
    setDecodedCopy((0, _decodeString.default)(copy));
  }, [copy]);
  /**
   * Render the Quote copy with the selected quote marks
   *
   * @returns {*} Blockquote with quote marks
   */

  var renderQuote = function renderQuote() {
    switch (markType) {
      case 'doubleCurved':
        return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
          className: "".concat(prefix, "--quote__mark")
        }, "\u201C"), /*#__PURE__*/_react.default.createElement("blockquote", {
          className: "".concat(prefix, "--quote__copy")
        }, decodedCopy, "\u201D"));

      case 'singleCurved':
        return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
          className: "".concat(prefix, "--quote__mark")
        }, "\u2018"), /*#__PURE__*/_react.default.createElement("blockquote", {
          className: "".concat(prefix, "--quote__copy")
        }, decodedCopy, "\u2019"));

      case 'doubleAngle':
        return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
          className: "".concat(prefix, "--quote__mark")
        }, "\xAB"), /*#__PURE__*/_react.default.createElement("blockquote", {
          className: "".concat(prefix, "--quote__copy")
        }, decodedCopy, "\xBB"));

      case 'singleAngle':
        return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
          className: "".concat(prefix, "--quote__mark")
        }, "\u2039"), /*#__PURE__*/_react.default.createElement("blockquote", {
          className: "".concat(prefix, "--quote__copy")
        }, decodedCopy, "\u203A"));

      case 'lowHighReversedDoubleCurved':
        return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
          className: "".concat(prefix, "--quote__mark")
        }, "\u201E"), /*#__PURE__*/_react.default.createElement("blockquote", {
          className: "".concat(prefix, "--quote__copy")
        }, decodedCopy, "\u201C"));

      case 'cornerBracket':
        return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
          className: "".concat(prefix, "--quote__mark ").concat(prefix, "--quote__mark-corner-bracket")
        }, "\u300C"), /*#__PURE__*/_react.default.createElement("blockquote", {
          className: "".concat(prefix, "--quote__copy")
        }, decodedCopy, "\u300D"));
    }
  };

  return /*#__PURE__*/_react.default.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--quote"),
    className: (0, _classnames2.default)("".concat(prefix, "--quote"), (0, _defineProperty2.default)({}, "".concat(prefix, "--quote__inverse"), inverse))
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--quote__container")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--quote__wrapper"),
    "data-autoid": "".concat(stablePrefix, "--quote__copy")
  }, renderQuote()), source && source.heading && source.copy ? /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--quote__source"),
    "data-autoid": "".concat(stablePrefix, "--quote__source")
  }, /*#__PURE__*/_react.default.createElement("p", {
    className: "".concat(prefix, "--quote__source-heading")
  }, source.heading), /*#__PURE__*/_react.default.createElement("p", {
    className: "".concat(prefix, "--quote__source-body")
  }, source.copy), source.copy2 && /*#__PURE__*/_react.default.createElement("p", {
    className: "".concat(prefix, "--quote__source-optional-copy")
  }, source.copy2)) : false), cta ? /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--quote__footer")
  }, /*#__PURE__*/_react.default.createElement(_HorizontalRule.HorizontalRule, null), /*#__PURE__*/_react.default.createElement(_LinkWithIcon.LinkWithIcon, {
    href: cta.href
  }, /*#__PURE__*/_react.default.createElement("span", null, cta.copy), cta.type === 'local' ? /*#__PURE__*/_react.default.createElement(_.default, null) : false, cta.type === 'external' ? /*#__PURE__*/_react.default.createElement(_2.default, null) : false)) : false);
};

Quote.propTypes = {
  /**
   * Type of the quote marks. Choose from:
   *
   * | Name                          | Description |
   * | ----------------------------- | ----------- |
   * | `singleCurved`                | ‘ ’         |
   * | `doubleCurved`                | “ ”         |
   * | `singleAngle`                 | ‹ ›         |
   * | `doubleAngle`                 | « »         |
   * | `lowHighReversedDoubleCurved` | „ “         |
   * | `cornerBracket`               |「 」         |
   */
  markType: _propTypes.default.oneOf(['doubleCurved', 'singleCurved', 'doubleAngle', 'singleAngle', 'lowHighReversedDoubleCurved', 'cornerBracket']),

  /**
   * Main Quote.
   */
  copy: _propTypes.default.string.isRequired,

  /**
   * Source object. The structure is:
   *
   * | Name      | Required | Data Type | Default Value | Description                  |
   * | --------- | -------- | --------- | ------------- | ---------------------------- |
   * | `heading` | YES      | String    | null          | Source heading               |
   * | `copy`    | YES      | String    | null          | Source body text             |
   * | `copy2`   | NO       | String    | null          | Optional second line of copy |
   */
  source: _propTypes.default.shape({
    heading: _propTypes.default.string.isRequired,
    copy: _propTypes.default.string.isRequired,
    copy2: _propTypes.default.string
  }),

  /**
   * Object with a sub-scheme of CTA data.
   * See the [`<CTA>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-cta--default#props) for full usage details.
   */
  cta: _propTypes.default.shape({
    copy: _propTypes.default.string,
    type: _propTypes.default.oneOf(['local', 'external']),
    href: _propTypes.default.string
  }),

  /**
   * `true` to use the invese colors.
   */
  inverse: _propTypes.default.bool
};
var _default = Quote;
exports.default = _default;
