// @carbon/ibmdotcom-react/lib/components/ContentBlockHorizontal/ContentBlockHorizontal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _ContentBlock = _interopRequireDefault(require("../ContentBlock/ContentBlock"));

var _ContentItemHorizontal = require("../ContentItemHorizontal");

var _settings = _interopRequireDefault(require("../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _settings2 = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var stablePrefix = _settings.default.stablePrefix;
var prefix = _settings2.default.prefix;
/**
 * ContentBlockHorizontal pattern.
 */

var ContentBlockHorizontal = function ContentBlockHorizontal(_ref) {
  var heading = _ref.heading,
      items = _ref.items,
      border = _ref.border;
  return /*#__PURE__*/_react.default.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--content-block-horizontal"),
    className: "".concat(prefix, "--content-block-horizontal")
  }, /*#__PURE__*/_react.default.createElement(_ContentBlock.default, {
    heading: heading,
    border: border
  }, items.map(function (item, index) {
    return /*#__PURE__*/_react.default.createElement(_ContentItemHorizontal.ContentItemHorizontal, {
      eyebrow: item.eyebrow,
      heading: item.heading,
      copy: item.copy,
      cta: item.cta,
      key: index
    });
  })));
};

ContentBlockHorizontal.propTypes = {
  /**
   * Heading of the content block.
   */
  heading: _propTypes.default.string.isRequired,

  /**
   * Array of content items.
   * See [`<ContentItemHorizontal>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-contentitemhorizontal--default#props) for full usage details.
   */
  items: _propTypes.default.arrayOf(_propTypes.default.shape({
    eyebrow: _propTypes.default.string,
    heading: _propTypes.default.string.isRequired,
    copy: _propTypes.default.string.isRequired,
    cta: _propTypes.default.shape({
      heading: _propTypes.default.string,
      items: _propTypes.default.arrayOf(_propTypes.default.shape({
        type: _propTypes.default.oneOfType([_propTypes.default.oneOf(['local', 'external']), _propTypes.default.arrayOf(_propTypes.default.oneOf(['local', 'external']))]),
        copy: _propTypes.default.string,
        href: _propTypes.default.string,
        customClassName: _propTypes.default.string
      })).isRequired,
      iconPlacement: _propTypes.default.oneOf(['left', 'right'])
    })
  })).isRequired,

  /**
   * border for content block.
   */
  border: _propTypes.default.bool
};
var _default = ContentBlockHorizontal;
exports.default = _default;
