// @carbon/ibmdotcom-services/es/services/Profile/Profile.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _regeneratorRuntime from "@babel/runtime/regenerator";

/**
 * Copyright IBM Corp. 2020, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import axios from 'axios';
/**
 * @constant {string | string} Host for the profile status API call
 * @private
 */

var _host = process && (process.env.REACT_APP_PROFILE_HOST || process.env.PROFILE_HOST) || 'https://login.ibm.com';
/**
 * @constant {string | string} API version
 * @private
 */


var _version = process && process.env.PROFILE_VERSION || 'v1';
/**
 * Profile status endpoint
 *
 * @type {string}
 * @private
 */


var _endpoint = "".concat(_host, "/").concat(_version, "/mgmt/idaas/user/status/");
/**
 * Profile API class with methods for checking user authentication for ibm.com
 */


var ProfileAPI = /*#__PURE__*/function () {
  function ProfileAPI() {
    _classCallCheck(this, ProfileAPI);
  }

  _createClass(ProfileAPI, null, [{
    key: "getUserStatus",

    /**
     * Returns user status (authenticated or unauthenticated)
     *
     * @returns {Promise<any>} User status
     * @example
     * import { ProfileAPI } from '@carbon/ibmdotcom-services';
     *
     * async function getUserStatus() {
     *   const response = await ProfileAPI.getUserStatus();
     *   return response;
     * }
     */
    value: function () {
      var _getUserStatus = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return axios.get(_endpoint, {
                  headers: {
                    'Content-Type': 'application/json; charset=utf-8'
                  },
                  withCredentials: true
                }).then(function (response) {
                  return response.data;
                }).catch(function (error) {
                  console.log('Failed Profile Network Call', error);
                  return {
                    user: 'Unauthenticated'
                  };
                });

              case 2:
                return _context.abrupt("return", _context.sent);

              case 3:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      function getUserStatus() {
        return _getUserStatus.apply(this, arguments);
      }

      return getUserStatus;
    }()
  }]);

  return ProfileAPI;
}();

export default ProfileAPI;
