// @carbon/ibmdotcom-react/lib/components/ContentBlock/ContentBlock.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _CTA = require("../CTA");

var _classnames = _interopRequireDefault(require("classnames"));

var _settings = _interopRequireDefault(require("../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings"));

var _HorizontalRule = require("../HorizontalRule");

var _markdownToHtml = _interopRequireDefault(require("@carbon/ibmdotcom-utilities/lib/utilities/markdownToHtml/markdownToHtml"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _settings2 = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var stablePrefix = _settings.default.stablePrefix;
var prefix = _settings2.default.prefix;
/**
 * ContentBlock internal component
 */

var ContentBlock = function ContentBlock(_ref) {
  var heading = _ref.heading,
      copy = _ref.copy,
      children = _ref.children,
      customClassName = _ref.customClassName,
      cta = _ref.cta,
      aside = _ref.aside,
      border = _ref.border;
  var classnames = (0, _classnames.default)("".concat(prefix, "--content-block"), customClassName);

  var content = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, copy && /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--content-block__copy"),
    dangerouslySetInnerHTML: {
      __html: (0, _markdownToHtml.default)(copy, {
        bold: false
      })
    }
  }), /*#__PURE__*/_react.default.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--content-block__children"),
    className: "".concat(prefix, "--content-block__children")
  }, children), cta && _renderCTA(cta, border));

  var title = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, heading && /*#__PURE__*/_react.default.createElement("h2", {
    "data-autoid": "".concat(stablePrefix, "--content-block__heading"),
    className: "".concat(prefix, "--content-block__heading")
  }, heading));

  return /*#__PURE__*/_react.default.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--content-block"),
    className: classnames
  }, title, aside && aside.items ? _layoutWrap(content, aside) : content, border ? /*#__PURE__*/_react.default.createElement(_HorizontalRule.HorizontalRule, null) : '');
};
/**
 * lays out the content with aside items in 2-1 format
 *
 * @private
 * @param {Element} content content elements
 * @param {Element} aside  aside elements
 * @returns {*} jsx cta component
 */


var _layoutWrap = function _layoutWrap(content, aside) {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--content-layout ").concat(prefix, "--content-layout--with-complementary")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--content-layout__body")
  }, content), /*#__PURE__*/_react.default.createElement("aside", {
    className: "".concat(prefix, "--content-layout__aside")
  }, aside.items));
};
/**
 * sets the class name based on theme type
 *
 * @private
 * @param {object} cta a cta object
 * @returns {*} jsx cta component
 */


function _renderCTA(cta, border) {
  if (cta.style === 'feature') {
    return /*#__PURE__*/_react.default.createElement(_CTA.CTA, (0, _extends2.default)({
      customClassName: (0, _classnames.default)("".concat(prefix, "--content-block__cta"))
    }, cta));
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--content-block__cta"),
    className: (0, _classnames.default)("".concat(prefix, "--content-block__cta-row"), (0, _defineProperty2.default)({}, "".concat(prefix, "--content-block__cta-row-border"), border))
  }, /*#__PURE__*/_react.default.createElement(_CTA.CTA, (0, _extends2.default)({
    customClassName: "".concat(prefix, "--content-block__cta ").concat(prefix, "--content-block__cta-col")
  }, cta)));
}

ContentBlock.propTypes = {
  /**
   * Heading text.
   */
  heading: _propTypes.default.string,

  /**
   * Copy text.
   */
  copy: _propTypes.default.string,

  /**
   * Children elements passed into `ContentBlock` to be rendered.
   */
  children: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.node), _propTypes.default.node]),

  /**
   * CTA object.
   * See the [`<CTA>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-cta--default#props) for full usage details.
   */
  cta: _propTypes.default.shape({
    style: _propTypes.default.oneOf(['text', 'card', 'button', 'feature']),
    type: _propTypes.default.oneOfType([_propTypes.default.oneOf(['jump', 'local', 'external', 'download', 'video']), _propTypes.default.arrayOf(_propTypes.default.oneOf(['jump', 'local', 'external', 'download', 'video']))]),
    copy: _propTypes.default.string,
    href: _propTypes.default.string,
    customClassName: _propTypes.default.string
  }),

  /**
   * Custom className to wrap the `<ContentBlock>` component.
   */
  customClassName: _propTypes.default.string,

  /**
   * Object containing elements to be rendered within `<aside>` html element on right panel.
   * The structure is:
   *
   * | Name     | Data Type | Description                                                |
   * | -------- | --------- | ---------------------------------------------------------- |
   * | `items`  | Element   | Elements/Components to be rendered on the right panel.     |
   */
  aside: _propTypes.default.shape({
    items: _propTypes.default.element
  }),

  /**
   * border for content block.
   */
  border: _propTypes.default.bool
};
ContentBlock.defaultProps = {
  border: false
};
var _default = ContentBlock;
exports.default = _default;
