// @carbon/ibmdotcom-react/es/components/carbon-components-react/UIShell/SideNavMenuSection.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";

/**
 * Copyright IBM Corp. 2016, 2021
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import ArrowRight20 from '@carbon/icons-react/es/arrow--right/20';
import cx from 'classnames';
import SideNavLink from '../../../internal/vendor/carbon-components-react/components/UIShell/SideNavLink';
import SideNavMenuBackButton from './SideNavMenuBackButton';
import settings from 'carbon-components/es/globals/js/settings';
import PropTypes from 'prop-types';
import React, { useRef, useEffect } from 'react';
var prefix = settings.prefix;
/**
 * SideNavMenuSection component
 */

var SideNavMenuSection = function SideNavMenuSection(_ref) {
  var _cx;

  var customClassName = _ref.className,
      children = _ref.children,
      onBackClick = _ref.onBackClick,
      show = _ref.show,
      rest = _objectWithoutProperties(_ref, ["className", "children", "onBackClick", "show"]);

  var menuSectionRef = useRef(null);
  var backButtonRef = useRef(null);
  useEffect(function () {
    if (show) {
      /**
       * In order for tabbing to work, focus has to be set in the menu section when it
       * becomes visible. If menu section is a submenu, set focus to the back button. If menu section
       * is the first parent section, set focus to the hamburger toggle button.
       *
       * @param {Node} focusElement node element to focus
       */
      var setFocus = function setFocus(focusElement) {
        var _menuSectionRef$curre;

        (_menuSectionRef$curre = menuSectionRef.current) === null || _menuSectionRef$curre === void 0 ? void 0 : _menuSectionRef$curre.addEventListener('transitionend', function focus(event) {
          var _menuSectionRef$curre2;

          if (event.propertyName === 'left' || event.propertyName === 'background-color' || event.propertyName === 'color') {
            focusElement === null || focusElement === void 0 ? void 0 : focusElement.focus();
          }

          (_menuSectionRef$curre2 = menuSectionRef.current) === null || _menuSectionRef$curre2 === void 0 ? void 0 : _menuSectionRef$curre2.removeEventListener('transitionend', focus);
        });
      };

      if (rest.isSubmenu) {
        setFocus(backButtonRef.current);
      } else {
        setFocus(rest.focusNode);
      }
    }
  }, [rest.focusNode, rest.isSubmenu, show]);
  var className = cx((_cx = {}, _defineProperty(_cx, "".concat(prefix, "--side-nav__menu-section"), true), _defineProperty(_cx, "".concat(prefix, "--side-nav__menu-section--expanded"), show), _defineProperty(_cx, customClassName, !!customClassName), _cx));
  return /*#__PURE__*/React.createElement("div", {
    "aria-hidden": !show,
    className: className,
    id: rest.id,
    ref: menuSectionRef
  }, rest.isSubmenu && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("button", {
    className: "".concat(prefix, "--masthead__focus"),
    onFocus: function onFocus(e) {
      var lastMenuItem = e.target.parentElement.querySelector('li:last-of-type button') || e.target.parentElement.querySelector('li:last-of-type a');
      return lastMenuItem.focus();
    },
    "aria-hidden": true
  }), /*#__PURE__*/React.createElement(SideNavMenuBackButton, {
    ref: backButtonRef,
    onBackClick: onBackClick,
    navType: rest.navType,
    backButtonText: rest.backButtonText
  }), rest.titleUrl ? /*#__PURE__*/React.createElement(SideNavLink, {
    className: "".concat(prefix, "--masthead__side-nav--submemu-section-title"),
    href: rest.titleUrl
  }, rest.title, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefix, "--masthead__side-nav--submemu-section-title__icon")
  }, /*#__PURE__*/React.createElement(ArrowRight20, null))) : /*#__PURE__*/React.createElement("li", {
    className: "".concat(prefix, "--masthead__side-nav--submemu-title")
  }, rest.title), rest.heading && /*#__PURE__*/React.createElement("p", {
    className: "".concat(prefix, "--masthead__side-nav--submemu-heading")
  }, rest.heading)), children, /*#__PURE__*/React.createElement("button", {
    className: "".concat(prefix, "--masthead__focus"),
    onFocus: function onFocus(e) {
      if (rest.focusNode) {
        rest.focusNode.focus();
      } else {
        e.target.parentElement.querySelector('button[isbackbutton="true"]').focus();
      }
    },
    "aria-hidden": true
  }));
};

SideNavMenuSection.propTypes = {
  /**
   * Provide an optional class to be applied to the containing node
   */
  className: PropTypes.string,

  /**
   * <SideNavMenuItem>'s within SideNavMenuSection
   */
  children: PropTypes.node,

  /**
   * action on back click
   */
  onBackClick: PropTypes.func,

  /**
   * Determine whether to show SideNavMenuSection
   */
  show: PropTypes.bool
};
SideNavMenuSection.defaultProps = {
  show: false,
  onBackClick: function onBackClick() {}
};
export default SideNavMenuSection;
