// @carbon/ibmdotcom-react/es/components/carbon-components-react/UIShell/SideNavMenuBackButton.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import ChevronLeft20 from '@carbon/icons-react/es/chevron--left/20';
import settings from 'carbon-components/es/globals/js/settings';
import ddsSettings from '../../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings';
import PropTypes from 'prop-types';
import React from 'react';
var stablePrefix = ddsSettings.stablePrefix;
var prefix = settings.prefix;
/**
 * SideNavMenu BackButton component
 */

var SideNavMenuBackButton = React.forwardRef(function SideNavMenuBackButton(props, ref) {
  var onBackClick = props.onBackClick,
      backButtonText = props.backButtonText,
      rest = _objectWithoutProperties(props, ["onBackClick", "backButtonText"]);

  var handleBackButtonClick = function handleBackButtonClick(event) {
    event.preventDefault();
    onBackClick();
  };

  var handleBackButtonKeyPress = function handleBackButtonKeyPress(event) {
    if (event.key === 'Enter' || event.charCode === ' ') {
      event.preventDefault();
      onBackClick();
    }
  };

  return /*#__PURE__*/React.createElement("li", {
    className: "".concat(prefix, "--side-nav__menu-item ").concat(prefix, "--masthead__side-nav--submemu-back")
  }, /*#__PURE__*/React.createElement("button", {
    ref: ref,
    onClick: handleBackButtonClick,
    onKeyPress: handleBackButtonKeyPress,
    className: "".concat(prefix, "--side-nav__link"),
    "data-autoid": "".concat(stablePrefix, "--masthead-").concat(rest.navType, "-sidenav__l0-back"),
    isbackbutton: "true"
  }, /*#__PURE__*/React.createElement("span", {
    className: "".concat(prefix, "--side-nav__link-text")
  }, /*#__PURE__*/React.createElement(ChevronLeft20, null), backButtonText)));
});
SideNavMenuBackButton.propTypes = {
  /**
   * action on back click
   */
  onBackClick: PropTypes.func,

  /**
   * back button text
   */
  backButtonText: PropTypes.string
};
SideNavMenuBackButton.defaultProps = {
  backButtonText: 'Back',
  onBackClick: function onBackClick() {}
};
export default SideNavMenuBackButton;
