// @carbon/ibmdotcom-react/es/internal/vendor/carbon-components-react/components/Grid/CSSGrid.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["as", "condensed", "fullWidth", "columns", "className", "children"];

/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import cx from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import { usePrefix } from '../../internal/usePrefix';
import { GridSettings, useGridSettings } from './GridContext';

function CSSGrid(_ref) {
  var _cx;

  var _ref$as = _ref.as,
      BaseComponent = _ref$as === void 0 ? 'div' : _ref$as,
      _ref$condensed = _ref.condensed,
      condensed = _ref$condensed === void 0 ? false : _ref$condensed,
      _ref$fullWidth = _ref.fullWidth,
      fullWidth = _ref$fullWidth === void 0 ? false : _ref$fullWidth,
      _ref$columns = _ref.columns,
      columns = _ref$columns === void 0 ? 16 : _ref$columns,
      containerClassName = _ref.className,
      children = _ref.children,
      rest = _objectWithoutProperties(_ref, _excluded);

  var prefix = usePrefix();

  var _useGridSettings = useGridSettings(),
      subgrid = _useGridSettings.subgrid;

  var className = cx(containerClassName, (_cx = {}, _defineProperty(_cx, "".concat(prefix, "--css-grid"), !subgrid), _defineProperty(_cx, "".concat(prefix, "--css-grid--").concat(columns), !subgrid && columns !== 16), _defineProperty(_cx, "".concat(prefix, "--css-grid--condensed"), condensed), _defineProperty(_cx, "".concat(prefix, "--css-grid--full-width"), fullWidth), _defineProperty(_cx, "".concat(prefix, "--subgrid"), subgrid), _defineProperty(_cx, "".concat(prefix, "--col-span-").concat(columns), subgrid && columns !== 16 || columns !== 16), _cx));
  return /*#__PURE__*/React.createElement(GridSettings, {
    mode: "css-grid",
    subgrid: true
  }, /*#__PURE__*/React.createElement(BaseComponent, _extends({
    className: className
  }, rest), children));
}

CSSGrid.propTypes = {
  /**
   * Provide a custom element to render instead of the default <div>
   */
  as: PropTypes.oneOfType([PropTypes.string, PropTypes.elementType]),

  /**
   * Pass in content that will be rendered within the `Grid`
   */
  children: PropTypes.node,

  /**
   * Specify a custom className to be applied to the `Grid`
   */
  className: PropTypes.string,

  /**
   * Specify how many columns wide the Grid should span
   */
  columns: PropTypes.number,

  /**
   * Collapse the gutter to 1px. Useful for fluid layouts.
   * Rows have 1px of margin between them to match gutter.
   */
  condensed: PropTypes.bool,

  /**
   * Remove the default max width that the grid has set
   */
  fullWidth: PropTypes.bool
};
export { CSSGrid };
