// @carbon/ibmdotcom-react/lib/components/LeadSpace/LeadSpace.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ButtonGroup = require("../../components/ButtonGroup");

var _classnames4 = _interopRequireDefault(require("classnames"));

var _settings = _interopRequireDefault(require("../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/settings/settings"));

var _deprecate = _interopRequireDefault(require("../../internal/vendor/@carbon/ibmdotcom-utilities/utilities/deprecate/deprecate.js"));

var _Image = require("../Image");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _settings2 = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2022
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var stablePrefix = _settings.default.stablePrefix;
var prefix = _settings2.default.prefix;
/**
 * renders main class name
 *
 * @param {string} type switches between centered or default
 * @param {string} theme theme of the pattern
 * @returns {string} classnames
 */

var classNames = function classNames(type, theme) {
  var _classnames;

  return (0, _classnames4.default)("".concat(prefix, "--leadspace__section"), (_classnames = {}, (0, _defineProperty2.default)(_classnames, "".concat(prefix, "--leadspace--").concat(theme), theme), (0, _defineProperty2.default)(_classnames, "".concat(prefix, "--leadspace--centered"), type === 'centered'), (0, _defineProperty2.default)(_classnames, "".concat(prefix, "--leadspace--productive"), type === 'small'), _classnames));
};
/**
 *
 * @param {object} image image
 * @returns {object} returns either image component or the centered image div
 */


function imageClassname(image) {
  return /*#__PURE__*/_react.default.createElement(_Image.Image, image);
}
/**
 * Lead space component (left-aligned)
 *
 * @param {object} props props object
 * @param {Array} props.buttons array of buttons for lead space (max 2 buttons)
 * @param {string} props.copy lead space short copy to support the title
 * @param {object} props.image image object with diff source for diff breakpoints
 * @param {string} props.theme theme of the pattern (g100 or white (default))
 * @param {string} props.title lead space title
 * @param {string} props.type type of lead space
 * @returns {*} Lead space component
 */


var LeadSpace = function LeadSpace(_ref) {
  var _classnames2;

  var buttons = _ref.buttons,
      copy = _ref.copy,
      image = _ref.image,
      theme = _ref.theme,
      title = _ref.title,
      type = _ref.type,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? 'tall' : _ref$size;
  return /*#__PURE__*/_react.default.createElement("div", {
    "data-autoid": "".concat(stablePrefix, "--leadspace"),
    className: (0, _classnames4.default)("".concat(prefix, "--leadspace"), (_classnames2 = {}, (0, _defineProperty2.default)(_classnames2, "".concat(prefix, "--leadspace--medium"), size === 'medium'), (0, _defineProperty2.default)(_classnames2, "".concat(prefix, "--leadspace--tall"), size === 'tall'), (0, _defineProperty2.default)(_classnames2, "".concat(prefix, "--leadspace--super"), size === 'super'), _classnames2))
  }, /*#__PURE__*/_react.default.createElement("section", {
    className: classNames(type, theme)
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--leadspace__container")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames4.default)("".concat(prefix, "--leadspace__overlay"), (0, _defineProperty2.default)({}, "".concat(prefix, "--leadspace--gradient"), image && image.defaultSrc))
  }, image && image.defaultSrc === true ? undefined : /*#__PURE__*/_react.default.createElement("svg", {
    className: "".concat(prefix, "--leadspace__gradient"),
    viewBox: "0 0 100 100",
    preserveAspectRatio: "none",
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink"
  }, /*#__PURE__*/_react.default.createElement("defs", null, /*#__PURE__*/_react.default.createElement("linearGradient", {
    id: "stops",
    className: "".concat(prefix, "--leadspace__gradient__stops"),
    gradientTransform: type === 'centered' ? 'rotate(90)' : ''
  }, type === 'centered' ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("stop", {
    offset: "0%"
  }), /*#__PURE__*/_react.default.createElement("stop", {
    offset: "54%"
  }), /*#__PURE__*/_react.default.createElement("stop", {
    offset: "77%"
  }), /*#__PURE__*/_react.default.createElement("stop", {
    offset: "100%"
  })) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("stop", {
    offset: "0%"
  }), /*#__PURE__*/_react.default.createElement("stop", {
    offset: "25%"
  }), /*#__PURE__*/_react.default.createElement("stop", {
    offset: "50%"
  }), /*#__PURE__*/_react.default.createElement("stop", {
    offset: "75%"
  })))), /*#__PURE__*/_react.default.createElement("rect", {
    className: "".concat(prefix, "--leadspace__gradient__rect"),
    width: "100",
    height: "100"
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--leadspace--content__container")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--leadspace__row")
  }, /*#__PURE__*/_react.default.createElement("h1", {
    className: "".concat(prefix, "--leadspace__title")
  }, title)), /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--leadspace__content")
  }, copy && /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefix, "--leadspace__row")
  }, /*#__PURE__*/_react.default.createElement("p", {
    "data-autoid": "".concat(stablePrefix, "--leadspace__desc"),
    className: "".concat(prefix, "--leadspace__desc")
  }, copy)), buttons && buttons.length > 0 && /*#__PURE__*/_react.default.createElement(_ButtonGroup.ButtonGroup, {
    buttons: buttons
  })))), image && imageClassname(image))));
};

LeadSpace.propTypes = {
  /**
   * Array of button objects to render (max 2).
   * See [`<ButtonGroup>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-buttongroup--default#button-item) for full usage details.
   */
  buttons: _propTypes.default.arrayOf(_propTypes.default.shape({
    copy: _propTypes.default.string.isRequired,
    href: _propTypes.default.string.isRequired,
    renderIcon: _propTypes.default.elementType
  })),

  /**
   * Short copy of LeadSpace.
   */
  copy: _propTypes.default.string,

  /**
   * Object with different ratio options for corresponding breakpoints.
   * See [`<Image>`'s README](http://www.ibm.com/standards/carbon/react/?path=/docs/components-image--default#props) for full usage details.
   */
  image: _propTypes.default.shape({
    classname: _propTypes.default.string,
    sources: _propTypes.default.arrayOf(_propTypes.default.shape({
      src: _propTypes.default.string,
      breakpoint: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number])
    })),
    defaultSrc: _propTypes.default.string.isRequired,
    alt: _propTypes.default.string.isRequired,
    longDescription: _propTypes.default.string
  }),

  /**
   * Color theme of LeadSpace. Choose from:
   *
   * | Name    | Data Type | Description           |
   * | ------- | --------- | --------------------- |
   * | `white` | String    | Carbon White theme    |
   * | `g10`   | String    | Carbon Gray 10 theme  |
   * | `g90`   | String    | Carbon Gray 90 theme  |
   * | `g100`  | String    | Carbon Gray 100 theme |
   */
  theme: _propTypes.default.oneOf(['white', 'g10', 'g90', 'g100']),

  /**
   * Title of LeadSpace.
   */
  title: _propTypes.default.string.isRequired,

  /**
   * Sets the type of Leadspace layout. Choose from:
   *
   * | Name              | Data Type | Description                                         |
   * | ----------------- | --------- | --------------------------------------------------- |
   * | `default`         | String    | Left-aligned - default style of the leadspace title |
   * | `left`            | String    | Left-aligned - large style of the leadspace title   |
   * | `centered`        | String    | Centered type of the LeadSpace                      |
   */
  type: _propTypes.default.oneOf(['default', 'left', 'centered']),

  /**
   * | Name         | Data Type | Description                           |
   * |--------------|-----------|---------------------------------------|
   * | `tall`/empty | String/-- | Default - tall size of the leadspace  |
   * | `medium`     | String    | Medium - medium size of the leadspace |
   * | `super`      | String    | Super - super size of the leadspace   |
   */
  size: _propTypes.default.oneOf(['tall', 'medium', 'super'])
};

var _default = (0, _deprecate.default)(LeadSpace, "\n  The Leadspace Small and Leadspace Small With Image variations are now deprecated.\n  Please refer to the Carbon for IBM.com documentation for further details.\n  https://www.ibm.com/standards/carbon/components/leadspace\n");

exports.default = _default;
